<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-gb"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Download larger amounts of genbank data into R — DoreyGenbank • Evolution Practicals</title><!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Download larger amounts of genbank data into R — DoreyGenbank"><meta name="description" content="This function works the same as rentrez::entrez_search() and ape::read.GenBank() together
to download data from Genbank. However, it overcomes issues of download limits by iteratively
downloading data and adding sequences and associated metadata together."><meta property="og:description" content="This function works the same as rentrez::entrez_search() and ape::read.GenBank() together
to download data from Genbank. However, it overcomes issues of download limits by iteratively
downloading data and adding sequences and associated metadata together."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Evolution Practicals</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html" aria-label="Go to package home"><span class="fa fa-home"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Available pages"><span class="fa fas fa-book"></span> Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/PhyloPrac1.html">Evolution Prac 1 — Getting ready!</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac2.html">Evolution Prac 2 — Access genetic data from GenBank</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac3.html">Evolution Prac 3 — Aligning genetic data and building phylogenies in R</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac4.html">Evolution prac 4 — Macroevolutionary questions!</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac5.html">Evolution prac 5 — Tips and pitfalls</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Download larger amounts of genbank data into R</h1>

      <div class="d-none name"><code>DoreyGenbank.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function works the same as <code><a href="https://docs.ropensci.org/rentrez/reference/entrez_search.html" class="external-link">rentrez::entrez_search()</a></code> and <code><a href="https://rdrr.io/pkg/ape/man/read.GenBank.html" class="external-link">ape::read.GenBank()</a></code> together
to download data from Genbank. However, it overcomes issues of download limits by iteratively
downloading data and adding sequences and associated metadata together.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">DoreyGenbank</span><span class="op">(</span></span>
<span>  db <span class="op">=</span> <span class="st">"nucleotide"</span>,</span>
<span>  term <span class="op">=</span> <span class="st">"(cytb[Gene Name]) AND (Apis[Organism])"</span>,</span>
<span>  retmax <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  config <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  retmode <span class="op">=</span> <span class="st">"xml"</span>,</span>
<span>  use_history <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  seq.names <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  species.names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  as.character <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  chunk.size <span class="op">=</span> <span class="fl">400</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"DNA"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-db">db<a class="anchor" aria-label="anchor" href="#arg-db"></a></dt>
<dd><p>character, name of the database to search for.</p></dd>


<dt id="arg-term">term<a class="anchor" aria-label="anchor" href="#arg-term"></a></dt>
<dd><p>character, the search term. The syntax used in making these searches is described in the Details of this help message, the package vignette and reference given below.</p></dd>


<dt id="arg-config">config<a class="anchor" aria-label="anchor" href="#arg-config"></a></dt>
<dd><p>vector configuration options passed to httr::GET</p></dd>


<dt id="arg-retmode">retmode<a class="anchor" aria-label="anchor" href="#arg-retmode"></a></dt>
<dd><p>character, one of json (default) or xml. This will make no difference in most cases.</p></dd>


<dt id="arg-use-history">use_history<a class="anchor" aria-label="anchor" href="#arg-use-history"></a></dt>
<dd><p>logical. If TRUE return a web_history object for use in later calls to the NCBI</p></dd>


<dt id="arg-seq-names">seq.names<a class="anchor" aria-label="anchor" href="#arg-seq-names"></a></dt>
<dd><p>the names to give to each sequence; by default the accession numbers are used. CURRENTLY NOT IN USE</p></dd>


<dt id="arg-species-names">species.names<a class="anchor" aria-label="anchor" href="#arg-species-names"></a></dt>
<dd><p>a logical indicating whether to attribute the species names to the returned object.</p></dd>


<dt id="arg-as-character">as.character<a class="anchor" aria-label="anchor" href="#arg-as-character"></a></dt>
<dd><p>a logical controlling whether to return the sequences as an object of class "DNAbin" (the default).</p></dd>


<dt id="arg-chunk-size">chunk.size<a class="anchor" aria-label="anchor" href="#arg-chunk-size"></a></dt>
<dd><p>the number of sequences downloaded together (see details).</p></dd>


<dt id="arg-quiet">quiet<a class="anchor" aria-label="anchor" href="#arg-quiet"></a></dt>
<dd><p>a logical value indicating whether to show the progress of the downloads. If TRUE, will also print the (full) name of the FASTA file containing the downloaded sequences.</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>a character specifying to download "DNA" (nucleotide) or "AA" (amino acid) sequences.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of DNA sequences made of vectors of class "DNAbin", or of single characters (if as.character = TRUE) with two attributes (species and description).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">TEST</span> <span class="op">&lt;-</span> <span class="fu">DoreyGenbank</span><span class="op">(</span>db <span class="op">=</span> <span class="st">"nucleotide"</span>,</span></span>
<span class="r-in"><span>term <span class="op">=</span> <span class="st">"(cytb[Gene Name]) AND (Apis[Organism])"</span>,</span></span>
<span class="r-in"><span>retmax <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by James B. Dorey.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

