<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Evolution Prac 3 — Aligning genetic data and building phylogenies in R • Evolution Practicals</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="../deps/MathJax-3.2.2/tex-chtml.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Evolution Prac 3 — Aligning genetic data and building phylogenies in R">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Evolution Practicals</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html" aria-label="Go to package home"><span class="fa fa-home"></span></a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Available pages"><span class="fa fas fa-book"></span> Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/PhyloPrac1.html">Evolution Prac 1 — Getting ready!</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac2.html">Evolution Prac 2 — Access genetic data from GenBank</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac3.html">Evolution Prac 3 — Aligning genetic data and building phylogenies in R</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac4.html">Evolution prac 4 — Macroevolutionary questions!</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac5.html">Evolution prac 5 — Tips and pitfalls</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">


<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Evolution Prac 3 — Aligning genetic data and building phylogenies in R</h1>
                        <h4 data-toc-skip class="author">Dr James B
Dorey and Dr Damien Esquerré, University of Wollongong</h4>
            
            <h4 data-toc-skip class="date">Version 2025-07-16; practical
2024-08-09</h4>
      

      <div class="d-none name"><code>PhyloPrac3.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      Today we will be re-downloading genetic data and starting to play
      around with making phylogenetic trees!<br><strong>Be certain to save your R script from this week to help
      you next week!</strong>
    </div>
    
<div class="section level2" number="1">
<h2 id="script-preparation">
<span class="header-section-number">1</span> Script preparation<a class="anchor" aria-label="anchor" href="#script-preparation"></a>
</h2>
<div class="section level3" number="1.1">
<h3 id="working-directory">
<span class="header-section-number">1.1</span> Working
directory<a class="anchor" aria-label="anchor" href="#working-directory"></a>
</h3>
<p>Just like last prac, we will set our working directory. This will,
likely, be the same directory as you used last week as it’s all one
project.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Set the RootPath to Prac1 folder</span></span>
<span><span class="va">RootPath</span> <span class="op">&lt;-</span> <span class="st">"YourFolderPathHere"</span></span>
<span>  <span class="co"># You can then set this as the project's working directory. </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="va">RootPath</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3" number="1.2">
<h3 id="install-packages">
<span class="header-section-number">1.2</span> Install packages<a class="anchor" aria-label="anchor" href="#install-packages"></a>
</h3>
<p>We should already have the packages from last week installed. If they
are not installed I’m sure that you can figure out how to do them using
the below example or copying the code from last prac!</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">packageList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="co"># Tidyverse packages:</span></span>
<span>  <span class="st">"tibble"</span>,            <span class="co">#  A lovely table format package</span></span>
<span>  <span class="st">"tidyr"</span>,             <span class="co"># Another data maniupulation package from the tidyverse</span></span>
<span>  <span class="st">"stringr"</span>           <span class="co">#  to manipulate text strings</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># List the new (not installed) packages and then if there are any, install them.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span>pkgs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">packageList</span><span class="op">)</span>, </span>
<span>              rebuild <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>              repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Media_EP3%2Ftidyr.png" width="100"><img src="Media_EP3%2Ftibble.png" width="100"><img src="Media_EP3%2Fstringr.png" width="100"></p>
</div>
<div class="section level3" number="1.3">
<h3 id="load-packages">
<span class="header-section-number">1.3</span> Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<p>Once again, we will need to load the packages into R, even the ones
that we used last week which were probably unloaded after you closed R
following our last session. This week, we’ll use a <em>base</em>
function, <code><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply()</a></code> that will <strong>apply</strong> a
function over a <strong>l</strong>ist. In this case, we will apply our
package list over the <code><a href="https://rdrr.io/r/base/library.html" class="external-link">library()</a></code> function. We will also add
in the relevant packages from last week that we want to use this week.
Do it all in one go!</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">packageList</span>, <span class="st">"dplyr"</span>, <span class="st">"magrittr"</span>, <span class="st">"ape"</span>, <span class="st">"rentrez"</span><span class="op">)</span>, </span>
<span>       <span class="va">library</span>, character.only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2" number="2">
<h2 id="genbank-data">
<span class="header-section-number">2</span> GenBank data<a class="anchor" aria-label="anchor" href="#genbank-data"></a>
</h2>
<div class="section level3" number="2.1">
<h3 id="read-data-in">
<span class="header-section-number">2.1</span> Read data in<a class="anchor" aria-label="anchor" href="#read-data-in"></a>
</h3>
<p>Let us start out by reading our fasta file backin from last week. If
you’ve not saved the file, go ahead and download my version from the
<strong>Moodle</strong> site and save it to your working directory
now.</p>
<p>In addition to last week I will also get an <strong>outgroup</strong>
sequence from a sister taxon, <em>Asemospiza obscura</em>, that will
help our phylogenetic analysis <strong>root</strong> our tree (i.e. it
will help place the most recent common ancestor for our finches). I
found this outgroup in the publication <a href="https://doi.org/10.1073/pnas.96.9.5101" class="external-link">“Phylogeny of Darwin’s
finches as revealed by mtDNA sequences”</a> and then found a cytB
sequence accession number on GenBank — “<strong>HQ153059</strong>”!</p>
<p><strong>Note:</strong> In this publication, the species was called
<em>Tiaris obscurus</em>, but it seems like the name has changed! This
is something to be aware of; taxonomy changes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in the data from last week</span></span>
<span><span class="va">DarwinFinches_cytB_lw</span> <span class="op">&lt;-</span> <span class="fu">base</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"DarwinFinches_cytB.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Download the outgroup cytB sequence, HQ153059, for the sister taxon, *Asemospiza obscura*</span></span>
<span><span class="va">A_obscura_cytB</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.GenBank.html" class="external-link">read.GenBank</a></span><span class="op">(</span>access.nb <span class="op">=</span> <span class="st">"HQ153059"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># We can combine the dataset from last week with the outgroup DNAbin objects simply</span></span>
<span><span class="co"># by appending them together.</span></span>
<span><span class="va">DarwinFinches_cytB</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB_lw</span>, <span class="va">A_obscura_cytB</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Annoyingly, we need to maually re-create the attributes for the DNAbin object and then re-</span></span>
<span><span class="co"># add them to the DNAbin object</span></span>
<span><span class="va">DarwinAttributes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB_lw</span><span class="op">)</span><span class="op">$</span><span class="va">names</span>, </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">A_obscura_cytB</span><span class="op">)</span><span class="op">$</span><span class="va">names</span><span class="op">)</span>,</span>
<span>  class <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB_lw</span><span class="op">)</span><span class="op">$</span><span class="va">class</span>,</span>
<span>  description <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB_lw</span><span class="op">)</span><span class="op">$</span><span class="va">description</span>, </span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">A_obscura_cytB</span><span class="op">)</span><span class="op">$</span><span class="va">description</span><span class="op">)</span>,</span>
<span>  species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB_lw</span><span class="op">)</span><span class="op">$</span><span class="va">species</span>, </span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">A_obscura_cytB</span><span class="op">)</span><span class="op">$</span><span class="va">species</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">DarwinAttributes</span></span></code></pre></div>
<p>Great, that was pretty easy wasn’t it? This demonstrates one of the
major advantages of running these kinds of things in a language like R;
once you have things set up you can run them again quite quickly. Not
only that, but what you did <strong>is easily reproducible and resilient
to new mistakes!</strong> That’s important in science.</p>
</div>
</div>
<div class="section level2" number="3">
<h2 id="filter-sequences">
<span class="header-section-number">3</span> Filter sequences<a class="anchor" aria-label="anchor" href="#filter-sequences"></a>
</h2>
<p>Alright, now that I’ve had my little preach, let’s address another
issue. We have 171 sequences, so there are more than one per species.
That might be fine depending on your analysis but it’ll complicate
things down the line in a way that we really just don’t need. Lets make
things easy; we’ll try to take <em>one individual per species</em>. But,
not all of the sequences are the same length! Longer sequences
<em>might</em> lead to better chances of overlapping our gene regions,
so lets first figure out how long each sequence is. Unfortunately,
<strong>ape</strong> doesn’t keep the sequence length in the attributes
of the file, but we can figure that out using <code><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply()</a></code> —
the same function that we used to read in all of our packages with the
<code><a href="https://rdrr.io/r/base/library.html" class="external-link">library()</a></code> function! Let us do the same with the
<code><a href="https://rdrr.io/r/base/length.html" class="external-link">length()</a></code> function.</p>
<p><strong>Note:</strong> Sometimes I’ll give you some complex/confusing
script, like the below, don’t stress too much about understanding these
data manipulations yet, just focus on the outputs and what they
mean.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Find the sequence lengths and turn them into a numeric list in order of how they were</span></span>
<span><span class="co"># found in DarwinFinches_cytB</span></span>
<span><span class="va">seqLengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB</span>, <span class="va">length</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p><strong>Q1:</strong> What is the range of sequence lengths in
seqLengths?</p>
</blockquote>
<p>Now, we know some lengths and we want to use those to help us choose
which sequences to take. What we’ll do now is create a data frame (or a
tibble from the <strong>tibble</strong> package which is nearly the same
thing but a bit nicer) with all of the information that we might want to
choose our sequences!</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Lets build a tibble column-by-column</span></span>
<span><span class="va">genBankSummaryTibble</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  accessionID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB</span><span class="op">)</span><span class="op">$</span><span class="va">names</span>,</span>
<span>  species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB</span><span class="op">)</span><span class="op">$</span><span class="va">species</span>,</span>
<span>  seqLength <span class="op">=</span> <span class="va">seqLengths</span>,</span>
<span>  description <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB</span><span class="op">)</span><span class="op">$</span><span class="va">description</span>,</span>
<span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Let's have a look!</span></span>
<span><span class="va">genBankSummaryTibble</span></span></code></pre></div>
<blockquote>
<p><strong>Q2:</strong> What do you think would happen in our table if
seqLengths were not in the same order of accession numbers as the rest
of the table?</p>
</blockquote>
<p>Great, now we can very quickly look at the data that we downloadeded
and make some choices on which sequences that we’re going to keep. For
simplicity’s sake, let’s just take the sequence that is longest per
species. To do that we will need to sort by sequence length and then
choose the top sequence. Who can help us with that? Our good friend
<strong>dplyr</strong>.</p>
<p><strong>Note:</strong> There may be some full or partial
mitochondrial genomes in your data (lengths &gt; 10,000 basepairs). We
will <code><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter()</a></code> these out for simplicity
<strong>Additionally</strong>, there are actually two sequences here (1)
the cytB sequence and (2) the control region for cytB. I’ll also remove
all of the “control region” sequences for you below; but be aware of
this for your project and check you alignments.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Let us begin to manipulate the tibble to do what we want!</span></span>
<span>  <span class="co"># We will put this new table into a new R object called theChosenSeqs</span></span>
<span><span class="va">theChosenSeqs</span> <span class="op">&lt;-</span> <span class="va">genBankSummaryTibble</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Keep only sequences less than 1,500 basepairs to ignore the mitogenomes</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seqLength</span> <span class="op">&lt;</span> <span class="fl">1500</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># remove the control region sequences</span></span>
<span>      <span class="co"># What's that "!" mean?! This simple bit of syntax just means "get the opposite of" </span></span>
<span>      <span class="co"># so, instead of returning all strigns where "control region" is detected, it </span></span>
<span>      <span class="co"># returns all where it is NOT detected.</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="va">description</span>, <span class="st">"control region"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Group the data by species and then any sorting or filtering will be done</span></span>
<span>    <span class="co"># WITHIN that group</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">species</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Sort by species and then by sequence length (top-down; descending)</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">species</span>, <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">seqLength</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Quickly, have a look at what you've made...</span></span>
<span><span class="va">theChosenSeqs</span></span>
<span></span>
<span><span class="co"># Now, let us take the top sequence of each group</span></span>
<span><span class="va">theChosenSeqs</span> <span class="op">&lt;-</span> <span class="va">theChosenSeqs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/row_number.html" class="external-link">row_number</a></span><span class="op">(</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p><strong>Q3:</strong> What’s the longest sequence and what’s the range
of sequence lengths?</p>
</blockquote>
<p>Amazing. So, now we have a table with the information that we want,
how do we filter our <em>DNAbin</em> object to only include the
specimens that we want? Well, now that we have found the longest
sequences using our data manipulation, let’s try and use the accession
numbers from the above table to filter our <em>DNAbin</em> object.</p>
<p>In R, there are usually several different ways to do the same thing.
But, we will use some <em>base</em> R syntax, the square brackets, the
will select the matching elements to what’s inside the square
brackets.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># List objects often have names for each element in the list; have a look at those</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># If we wanted to take the FIRST element, we could simply call the first element</span></span>
<span><span class="va">DarwinFinches_cytB</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co"># We could call an element using its name</span></span>
<span><span class="va">DarwinFinches_cytB</span><span class="op">[</span><span class="st">"56682246"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Extending this logic, we can use the accessionID column from theChosenSeqs to filter out</span></span>
<span><span class="co"># all of the sequences that we're interested in.</span></span>
<span><span class="va">finchSpecies</span> <span class="op">&lt;-</span> <span class="va">DarwinFinches_cytB</span><span class="op">[</span><span class="va">theChosenSeqs</span><span class="op">$</span><span class="va">accessionID</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># But you know what having accession numbers alone is a bit annoying. Lets add</span></span>
<span><span class="co"># species names as well!</span></span>
<span>  <span class="co"># NOTE: This is an example of complex code that you need not worry too much about</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">finchSpecies</span><span class="op">)</span><span class="op">$</span><span class="va">names</span> <span class="op">&lt;-</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_c.html" class="external-link">str_c</a></span><span class="op">(</span><span class="va">theChosenSeqs</span><span class="op">$</span><span class="va">species</span>,</span>
<span>                                                 <span class="va">theChosenSeqs</span><span class="op">$</span><span class="va">accessionID</span>,</span>
<span>                                                 sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p><strong>Q4:</strong> What is the name of the 5th sequence in
DarwinFinches_cytB?</p>
</blockquote>
<blockquote>
<p><strong>Q5:</strong> What is the name of the 5th sequence in
finchSpecies?</p>
</blockquote>
<blockquote>
<p><strong>Q6</strong> If you run “finchSpecies”, what is the base
composition (proportion) of “a”?</p>
</blockquote>
</div>
<div class="section level2" number="4">
<h2 id="align-sequences">
<span class="header-section-number">4</span> Align sequences<a class="anchor" aria-label="anchor" href="#align-sequences"></a>
</h2>
<p>Now we have 15 sequences; one per species in the cytB dataset on
GenBank. But, the sequences are so far just running from their first
base pair until their last… We can’t say that each base pair at position
1 are the same by descent (that they are homologous). We want to compare
how base pairs might have changed over time at each and every site along
the sequences; we need to <strong>align the sequences</strong>.
Sometimes this is straightforwards, sometimes it really is not. But
let’s find out together if we’ve got an easy alignment ahead of us!</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Here, we'll use a package called msaR to observe and align our sequences</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"msaR"</span>,</span>
<span>              repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span>
<span><span class="co"># We also need the BiocManager to install packages from bioconductor, where many </span></span>
<span><span class="co"># genetics packages, like Biostrings and msa, are found</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span>,</span>
<span>              repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span>
<span>  <span class="co"># you may get asked to install extra packages, select all "a"</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"Biostrings"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"msa"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">msaR</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/UBod/msa" class="external-link">msa</a></span><span class="op">)</span></span></code></pre></div>
<p><img src="Media_EP3%2FBiocManager.png" width="100"><img src="Media_EP3%2FmsaR.png" width="100"><img src="Media_EP3%2FBiostrings.png" width="100"><img src="Media_EP3%2Fmsa.png" width="100"></p>
<p><br></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># msaR agrees that fasta is a fantastic data format and but it recognises that not </span></span>
<span><span class="co"># everyone thinks this way. So, it has provided a function to transform a DNAbin object,</span></span>
<span><span class="co"># among several others, to fasta</span></span>
<span><span class="va">finchSpecies_fasta</span> <span class="op">&lt;-</span> <span class="fu">msaR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/msaR/man/as.fasta.html" class="external-link">as.fasta</a></span><span class="op">(</span><span class="va">finchSpecies</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># You can have a look at this output file and it looks like a horrible text string, but</span></span>
<span><span class="co"># if you look closer, you'll see that it's a fasta text string with line breaks denoted</span></span>
<span><span class="co"># by "\n" for "new line".</span></span>
<span><span class="va">finchSpecies_fasta</span></span>
<span></span>
<span><span class="co"># But, many R programs demand sequences of equal length. Don't worry too much about the </span></span>
<span><span class="co"># below but let me help you out here...</span></span>
<span>  <span class="co"># Please DON'T FREAK OUT when seeing this code; I'm just giving it to you </span></span>
<span>  <span class="co"># to do some data wizardry (with some notes IF you're interested)</span></span>
<span><span class="co"># First, lets add our fasta text into the "sequence" column of a new tibble (data frame)</span></span>
<span><span class="va">lengthCorrectionTable</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>sequence <span class="op">=</span> <span class="va">finchSpecies_fasta</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Separate sequences from one another, using the "&gt;" as a delimiter </span></span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate_longer_delim.html" class="external-link">separate_longer_delim</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="st">"sequence"</span>,</span>
<span>                               delim <span class="op">=</span> <span class="st">"&gt;"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="co"># Remove the first row from the tibble</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/row_number.html" class="external-link">row_number</a></span><span class="op">(</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1</span> <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Remove the hanging line break at the end of each sequence</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>sequence <span class="op">=</span> <span class="va">sequence</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="st">"\\n$"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="co"># Use the name &gt; sequence line break to separate them</span></span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate_wider_delim.html" class="external-link">separate_wider_delim</a></span><span class="op">(</span><span class="va">sequence</span>, delim <span class="op">=</span> <span class="st">"\n"</span>,</span>
<span>                              names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"name"</span>,<span class="st">"sequence"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="co"># THIS is the magic part, let's add blank "-" loci to pad out our alignments to the same length</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>lengthenedSequence <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_pad.html" class="external-link">str_pad</a></span><span class="op">(</span><span class="va">sequence</span>, pad <span class="op">=</span> <span class="st">"-"</span>, side <span class="op">=</span> <span class="st">"right"</span>,</span>
<span>                                                      width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_count.html" class="external-link">str_count</a></span><span class="op">(</span><span class="va">sequence</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                  <span class="co"># If you want, you can check this object to see that lengths are the same</span></span>
<span>                length2 <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_count.html" class="external-link">str_count</a></span><span class="op">(</span><span class="va">lengthenedSequence</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span><span class="co"># Now, we can turn this back into a fasta format... Asian don't worry too much about</span></span>
<span><span class="co"># this code.</span></span>
<span><span class="va">equalisedFasta</span> <span class="op">&lt;-</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_c.html" class="external-link">str_c</a></span><span class="op">(</span><span class="st">"&gt;"</span>,<span class="va">lengthCorrectionTable</span><span class="op">$</span><span class="va">name</span>,<span class="st">"\n"</span>,</span>
<span>                                 <span class="va">lengthCorrectionTable</span><span class="op">$</span><span class="va">lengthenedSequence</span>, </span>
<span>                                 collapse <span class="op">=</span> <span class="st">"\n"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Now, let's save this fasta file for later use</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/write.html" class="external-link">write</a></span><span class="op">(</span><span class="va">equalisedFasta</span>, <span class="st">"equalisedFasta.fasta"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># We can read this fasta in as a Biostrings object</span></span>
<span><span class="va">equalised_BioString</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="st">"equalisedFasta.fasta"</span><span class="op">)</span></span></code></pre></div>
<p>If you zoned out for that last section, start paying attention again
here!</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># START PAYING ATTENTION AGAIN!</span></span>
<span><span class="co"># finally, we can interactively look at our alignment!</span></span>
<span><span class="fu">msaR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/msaR/man/msaR.html" class="external-link">msaR</a></span><span class="op">(</span><span class="va">equalised_BioString</span>, menu<span class="op">=</span><span class="cn">T</span>, overviewbox <span class="op">=</span> <span class="cn">T</span>, labelNameLength <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<p>Take a deep breath, that was quite an adventure! But now we have had
a first look at our sequences. You’ll notice that things look a little
messy, especially that the letters don’t really line up nicely along
columns. You should expect this for an <strong>unaligned</strong> set of
sequences. Maybe you can even start to see where they might match up!
So, let’s see if we can change that.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We can use the msa, Multiple Sequence Alignment package to align our sequences </span></span>
<span><span class="va">alignedFinches</span> <span class="op">&lt;-</span> <span class="fu">msa</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/msa/man/msa.html" class="external-link">msa</a></span><span class="op">(</span><span class="va">equalised_BioString</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Then we can access the @unmasked slot of the alignment and view the results</span></span>
<span><span class="fu">msaR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/msaR/man/msaR.html" class="external-link">msaR</a></span><span class="op">(</span><span class="va">alignedFinches</span><span class="op">@</span><span class="va">unmasked</span>, menu<span class="op">=</span><span class="cn">T</span>, overviewbox <span class="op">=</span> <span class="cn">T</span>, labelNameLength <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<p>How does this alignment look? Can you see sections that align better
than others? Do you see sequences that look better than others? In some
cases, there may be a sequence that does not align well and may need to
be removed. For your project, if this is a problem, you may choose to
remove it, and add a different sequence in for that species. You can
feel free to do that either in a text editor or in R; whatever you’re
most comfortable with.</p>
<p>Now we have a reasonable alignment of 15 species. That’s certainly
enough to make a tree with!</p>
</div>
<div class="section level2" number="5">
<h2 id="make-a-phylogeny">
<span class="header-section-number">5</span> Make a phylogeny<a class="anchor" aria-label="anchor" href="#make-a-phylogeny"></a>
</h2>
<p>As you will learn in your next lecture, there are a few different
ways to build trees. But, for now let’s keep it pretty simple.</p>
<div class="section level3" number="5.1">
<h3 id="distance-tree">
<span class="header-section-number">5.1</span> Distance tree<a class="anchor" aria-label="anchor" href="#distance-tree"></a>
</h3>
<p>Lets make one of the simplest trees possible, a
<em>neighbour-joining</em> tree. This kind of tree is built based on
genetic distances between sequences. The closer the sequences, the
closer they’ll be on the tree! We can also tell <strong>ape</strong>
that we have an outgroup species and it will make that species sister to
all other taxa.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Let us start by converting the msa format sequences back into ape's DNAbin format</span></span>
<span><span class="va">alignedDNAbin</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/as.alignment.html" class="external-link">as.DNAbin</a></span><span class="op">(</span><span class="va">alignedFinches</span><span class="op">@</span><span class="va">unmasked</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># We then need to build a pairwise distance matrix — the distances between each</span></span>
<span><span class="co"># pair of sequences </span></span>
<span><span class="va">distanceFinches</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/dist.dna.html" class="external-link">dist.dna</a></span><span class="op">(</span><span class="va">alignedDNAbin</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Go ahead and have a look at the matrix</span></span>
<span><span class="va">distanceFinches</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Now we can use that distance matrix to build a phylogenetic tree</span></span>
<span><span class="va">njFinches</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/nj.html" class="external-link">nj</a></span><span class="op">(</span><span class="va">distanceFinches</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Specify the outgroup to be sister to all other taxa</span></span>
<span>  <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/root.html" class="external-link">root</a></span><span class="op">(</span>outgroup <span class="op">=</span> <span class="st">"Asemospiza_obscura_HQ153059"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># This is a tree object of class "phylo", and we need to visualise it with another </span></span>
<span><span class="co"># function... creatively called "plot"</span></span>
<span><span class="va">njFinches</span></span>
<span></span>
<span><span class="co"># Let's have a look!</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">njFinches</span><span class="op">)</span></span>
<span>  <span class="co"># If you have gaps in your sequence, you could look at ?ape::njs()</span></span>
<span></span>
<span><span class="co"># the ape package also provides an "improved" version of the nj algorithm. </span></span>
<span>  <span class="co"># Why not compare the tree?</span></span>
<span><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/bionj.html" class="external-link">bionj</a></span><span class="op">(</span><span class="va">distanceFinches</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/root.html" class="external-link">root</a></span><span class="op">(</span>outgroup <span class="op">=</span> <span class="st">"Asemospiza_obscura_HQ153059"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># Are you starting to understand how we use the pipes (%&gt;%) from the magrittr package?</span></span></code></pre></div>
<blockquote>
<p><strong>Q7:</strong> What instantly jumps out about the nj tree and
the groupings?</p>
</blockquote>
<blockquote>
<p><strong>Q8:</strong> Are the nj and bionj trees very different?</p>
</blockquote>
</div>
<div class="section level3" number="5.2">
<h3 id="maximum-likelihood">
<span class="header-section-number">5.2</span> Maximum
likelihood<a class="anchor" aria-label="anchor" href="#maximum-likelihood"></a>
</h3>
<p>The <em>neighbour-joining</em> tree is nice, quick, easy, but not the
most robust implementation for tree-building. Other methods exist which
you will learn more about next week. But, for now let’s have a peak at
these methods and the trees and see if anything jumps out at us. We’ll
want yet another package, for this — “<em>phangorn</em>”, Phylogenetic
Reconstruction and Analysis.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install and load phangorn</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"phangorn"</span>,</span>
<span>              repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KlausVigo/phangorn" class="external-link">phangorn</a></span><span class="op">)</span></span></code></pre></div>
<p><img src="Media_EP3%2Fphangorn.png" width="100"></p>
<p><br></p>
<p>Next week we will be talking about genealogies and models of
evolution so I don’t want to overload you now. However, a short prelude
to that might be: A, T, C, and G can each evolve differently and can
mutate to-and-from different states in a way that we would
<strong>not</strong> expect from random chance. So, clever folks have
come up with various <em>models of evolution</em> to account for this.
The model that we choose depends on our data in terms of which model
best fits our data and if our data are rich enough to make using a more
complex model possible! Let us use <strong>phangorn</strong>’s
<code><a href="https://klausvigo.github.io/phangorn/reference/modelTest.html" class="external-link">modelTest()</a></code> function to look at this.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Sadly we need to transform yet another data format from DNAbin to phyDat</span></span>
<span><span class="co"># Don't ask me why they do this to us ;)</span></span>
<span><span class="va">aligned_phyDat</span> <span class="op">&lt;-</span> <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/as.phyDat.html" class="external-link">phyDat</a></span><span class="op">(</span><span class="va">alignedDNAbin</span><span class="op">)</span></span>
<span>  <span class="co"># Do you see how we try to name objects so that they are concise but human-readable?</span></span>
<span></span>
<span><span class="co"># Run the model test</span></span>
<span><span class="va">finchTest</span> <span class="op">&lt;-</span> <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/modelTest.html" class="external-link">modelTest</a></span><span class="op">(</span><span class="va">aligned_phyDat</span>,</span>
<span>                              <span class="co"># Which models to compare</span></span>
<span>                            model<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"JC"</span>, <span class="st">"F81"</span>, <span class="st">"K80"</span>, <span class="st">"HKY"</span>, <span class="st">"SYM"</span><span class="op">)</span>,</span>
<span>                              <span class="co"># Let's not worry about invariant (I) or gamma (G)</span></span>
<span>                              <span class="co"># variables for now. Look these up later IF you need</span></span>
<span>                            I <span class="op">=</span> <span class="cn">FALSE</span>, G <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </span></code></pre></div>
<p>The <code><a href="https://klausvigo.github.io/phangorn/reference/modelTest.html" class="external-link">modelTest()</a></code> function lets you compare which model is
the best fit for the data. Let me help you at least interpret the values
that it returns! I don’t expect you to remember these, but they are
helpful concepts for phylogenetics and statistics at large.</p>
<ul>
<li>
<strong>df</strong> — the degrees of freedom</li>
<li>
<strong>logLike</strong> — the log of the likelihood. Because
likelihood values are often very small (like <strong>really</strong>
small), we take their log. Log values &lt;1 are negative values and they
get more negative as you add decimal places. Hence, the larger number
(e.g., -100 &gt; -500) is more likely! <strong>Side quest:</strong> why
don’t you compare the outputs of <em>log(0.1)</em>;
<em>log(0.0001)</em>; and <em>log(0.00000001)</em>
</li>
<li>
<strong>AIC</strong> — the Akaike information criterion provides a
comparison between models, given a set of data. Again, a smaller AIC
value represents a more <em>parsimonious</em> (i.e., simpler) option,
given the data. As a relative measure of support these numbers can only
be compared within alternative hypotheses.</li>
<li>
<strong>BIC</strong> — the Bayesian Information Criterion is very
similar to the AIC except it also considers the number of observations
in the formula. But again, a lower value is better.</li>
</ul>
<blockquote>
<p><strong>Q9:</strong> Which model is best supported by all three
statistics?</p>
</blockquote>
<p>You can also ask phangorn to choose the best model for you!</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Find me the best model, phangoorn!</span></span>
<span><span class="va">fit_mt</span> <span class="op">&lt;-</span> <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/pml_bb.html" class="external-link">pml_bb</a></span><span class="op">(</span><span class="va">finchTest</span>, </span>
<span>                  <span class="co"># Change to trace = 1 to see more outputs</span></span>
<span>                 control <span class="op">=</span> <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/pml.control.html" class="external-link">pml.control</a></span><span class="op">(</span>trace <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Now have a look at the best model and the relative rates of mutations table! </span></span>
<span><span class="va">fit_mt</span></span></code></pre></div>
<blockquote>
<p><strong>Q10:</strong> How many different rates are there REALLY in
this model?</p>
</blockquote>
<blockquote>
<p><strong>Q11:</strong> What are the relative rates of change for these
rates?</p>
</blockquote>
<p>Now, let’s make a bootstrapped <em>maximum likelihood</em> tree!</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># You can then feed in the fit_mt from above to build your bootstrap tree</span></span>
<span><span class="va">bs</span> <span class="op">&lt;-</span> <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/bootstrap.pml.html" class="external-link">bootstrap.pml</a></span><span class="op">(</span><span class="va">fit_mt</span>, </span>
<span>                                <span class="co"># Using 100 bootstraps</span></span>
<span>                              bs <span class="op">=</span> <span class="fl">100</span>, optNni <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                              method <span class="op">=</span> <span class="st">"ultrametric"</span>,</span>
<span>                              control <span class="op">=</span> <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/pml.control.html" class="external-link">pml.control</a></span><span class="op">(</span>trace <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot your maximum likelihood tree. You will input a "tree" which should be the</span></span>
<span><span class="co"># "Maximum clade credibility" tree (i.e., the statistically best tree from your </span></span>
<span><span class="co"># bootstrap), then you can give the bootstrap trees (BStrees), which include ALL</span></span>
<span><span class="co"># of the bootstrap trees</span></span>
<span><span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/plotBS.html" class="external-link">plotBS</a></span><span class="op">(</span></span>
<span>  tree <span class="op">=</span> <span class="va">bs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>      <span class="co"># Find the Maximum clade credibility tree</span></span>
<span>    <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/maxCladeCred.html" class="external-link">maxCladeCred</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>      <span class="co"># We can root the tree at this point</span></span>
<span>    <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/root.html" class="external-link">root</a></span><span class="op">(</span>outgroup <span class="op">=</span> <span class="st">"Asemospiza_obscura_HQ153059"</span><span class="op">)</span>,</span>
<span>  BStrees <span class="op">=</span> <span class="va">bs</span> , </span>
<span>  p <span class="op">=</span> <span class="fl">10</span>, type<span class="op">=</span><span class="st">"p"</span>, digits<span class="op">=</span><span class="fl">2</span>, main<span class="op">=</span><span class="st">"Ultrafast bootstrap"</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p><strong>Q12:</strong> What’s different about this tree from the last
one?</p>
</blockquote>
<blockquote>
<p><strong>Q13:</strong> Are some nodes better supported than
others?</p>
</blockquote>
<p>I am also going to save the DNAbin data for use in next week’s
prac.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Save the DNAbin as a fasta file</span></span>
<span><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/write.dna.html" class="external-link">write.FASTA</a></span><span class="op">(</span><span class="va">alignedDNAbin</span>,</span>
<span>                 <span class="st">"alignedDNAbin.fasta"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2" number="6">
<h2 id="whats-coming-next-week">
<span class="header-section-number">6</span> What’s coming next
week?<a class="anchor" aria-label="anchor" href="#whats-coming-next-week"></a>
</h2>
<p>Next week we will be focusing on doing something interesting and
useful with our phylogeny! We will be starting to think about harvesting
ecological, phenotypic, or biogeographic data to test a
macro-evolutionary hypothesis using your phylogeny. We will be using
Darwin’s finches once again but you should be advanced enough to have at
the very least an idea of what group you will examine for your
project.</p>
<p><strong>Be certain to save your R script from this week to help you
next week!</strong> If you don’t, well… sounds like you’ve got some
catching up to do.</p>
<p><br><br><br><br></p>
</div>
<div class="section level2" number="7">
<h2 id="new-packages-used-today">
<span class="header-section-number">7</span> New packages used
today<a class="anchor" aria-label="anchor" href="#new-packages-used-today"></a>
</h2>
<p><img src="Media_EP3%2Ftidyr.png" width="200"><img src="Media_EP3%2Ftibble.png" width="200"><img src="Media_EP3%2Fstringr.png" width="200"><img src="Media_EP3%2FBiocManager.png" width="200"><img src="Media_EP3%2FmsaR.png" width="200"><img src="Media_EP3%2FBiostrings.png" width="200"><img src="Media_EP3%2Fmsa.png" width="200"><img src="Media_EP3%2Fphangorn.png" width="200"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James B. Dorey.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
