<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-gb">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Evolution Prac MEGA supplement — Using MEGA to align sequences and build trees • Evolution Practicals</title>
<!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Evolution Prac MEGA supplement — Using MEGA to align sequences and build trees">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Evolution Practicals</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html" aria-label="Go to package home"><span class="fa fa-home"></span></a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Available pages"><span class="fa fas fa-book"></span> Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/PhyloPrac1.html">Evolution Prac 1 — Getting ready!</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac2.html">Evolution Prac 2 — Access genetic data from GenBank</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac3.html">Evolution Prac 3 — Aligning genetic data and building phylogenies in R</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac4.html">Evolution prac 4 — Macroevolutionary questions!</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac5.html">Evolution prac 5 — Tips and pitfalls</a></li>
    <li><a class="dropdown-item" href="../articles/MEGA_tutorial.html">Evolution Prac MEGA supplement — Using MEGA to align sequences and build trees</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">


<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Evolution Prac MEGA supplement — Using MEGA to align sequences and build trees</h1>
                        <h4 data-toc-skip class="author">Dr James B
Dorey and Dr Damien Esquerré, University of Wollongong</h4>
            
            <h4 data-toc-skip class="date">Version 2025-07-18; practical
NA</h4>
      

      <div class="d-none name"><code>MEGA_tutorial.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      This is a supplementary practical to show you how to align
      sequences in MEGA and also make trees! Wow!<br><strong>Be certain to save your data and take good notes on what
      you’ve done!</strong>
    </div>
    
<div class="alert alert-danger" role="alert">
<p><strong>Science tip:</strong> Using GUI (or Graphical User Interface)
programs is quite normal in scientific research. Many researchers even
prefer them! For myself, I have a long history of making phylogenies
using a program called <strong>BEAST2</strong>, which is a GUI program.
Thankfully, many of these programs create files that run code in the
background that can be read by parts of that program or by the user to
<em>remember what the heck they did!</em></p>
<p>That might seem silly, but often we are quickly running analyses and
doing things and then some months — or years — down the track we need to
write EXACTLY what we did in the methods sections of our scientific
manuscripts! With <strong>R</strong> coding and some other packages, so
long as you save your data and know the version that you used, all of
this is quite clear and can be referred back to easily enough. With GUI
programs that may not always be the case.</p>
<p>My advice then? Take excellent notes of what you did when using a
GUI. Even when I use some other coding languages (especially
<strong>bash</strong> in Mac’s <strong>Terminal</strong>) I will take
notes of exactly what I did so that when I come back to it later, I have
a chance of explaining myself or finding errors!</p>
<p>That’s a long-winded way of saying and justifying that you need to
take good notes. Please. For all of our sake’s.</p>
</div>
<div class="section level2">
<h2 id="intro-to-mega">Intro to <strong>MEGA</strong><a class="anchor" aria-label="anchor" href="#intro-to-mega"></a>
</h2>
<p>Did you know that I’ve never actually used MEGA before? Ha! So,
welcome to my tutorial where I teach you as I learn! <em>I sure know how
ot inspire confidence.</em> But, I have used a similar, but paid program
called <strong>Geneious</strong> quite extensively. MEGA, <a href="https://www.megasoftware.net" class="external-link">downloadable here</a>, is a powerful
free software that can really do a variety of things. It has been around
for a long time and is still being maintained and updated, which is
amazing!</p>
<div class="section level3">
<h3 id="downloading-mega">Downloading <strong>MEGA</strong><a class="anchor" aria-label="anchor" href="#downloading-mega"></a>
</h3>
<p>From the <a href="https://www.megasoftware.net" class="external-link">MEGA home page</a>
you can download the program. Note, that it has drop-downs to help you
choose the operating system, version (GUI or command-line [code]), and
version that you want. But, this should automatically populate correctly
for you. Go ahead and click the green “<strong>DOWNLOAD</strong>”
button.</p>
<p><img src="Media_MEGA/Mega_download1.png"></p>
<p>You will then need to scroll down and “<strong>Accept</strong>” the
user agreement.</p>
<p>Then, you will need to fill out a short survey about who you are
(similar to below). <strong>Note:</strong> that <strong>MEGA</strong> is
free for <em>research</em> and <em>education</em>. So, make sure that
you tick at least one of those :) Then, you will be able to use the
download client to install it on your computer.</p>
<p><img src="Media_MEGA/Mega_download2.png"></p>
</div>
<div class="section level3">
<h3 id="running-mega">Running <strong>MEGA</strong><a class="anchor" aria-label="anchor" href="#running-mega"></a>
</h3>
<p>When you first open <strong>MEGA</strong>, you will likely get asked
some questions about yourself… Feel free to answer, or not. That’s up to
you, but generally speaking, when someone makes resources freely
available, I’m quite inclined to help out and answer!</p>
<p><img src="Media_MEGA/Mega_download3.png"></p>
<p>Take a moment to look at <strong>MEGA</strong> and some of the
options it provides. It also provides help for first-time users and
tutorials (as below) that you might find useful!</p>
<p><img src="Media_MEGA/MEGA_help1.png"></p>
<p>This includes tutorials on aligning sequences and whatever else you
could want — hey, this is a pretty cool program!</p>
<p><img src="Media_MEGA/MEGA_help2.png"></p>
</div>
<div class="section level3">
<h3 id="open-your-data">Open your data<a class="anchor" aria-label="anchor" href="#open-your-data"></a>
</h3>
<p>First thing’s first, let’s open our alignment in
<strong>MEGA</strong>. You can click on the “<strong>DATA</strong>” and
then navigate to the alignment you want to use. I will use our
<em>equalisedFasta.fasta</em> file, which you can download using the
below code in <strong>R</strong> (or by pasting that link into your
browser and copying the text).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/jbdorey/EvolutionPracs/main/vignettes/equalisedFasta.fasta"</span>,</span>
<span>                     destfile <span class="op">=</span> <span class="st">"equalisedFasta.fasta"</span>,</span>
<span>                     method<span class="op">=</span><span class="st">"curl"</span><span class="op">)</span></span></code></pre></div>
<p>You will then be asked if you’d like to <em>Align</em> or
<em>Analyze</em> your data import. Choose <strong>Align</strong>.</p>
<ol style="list-style-type: decimal">
<li><img src="Media_MEGA/OpenData.png"></li>
<li><img src="Media_MEGA/OpenData2.png"></li>
<li><img src="Media_MEGA/OpenData3.png"></li>
</ol>
<p>The alignment that opens up should look pretty similar to the one
that I helped you to make in practical 3! We can see the
species/sequence names in a column down the left-hand side and we can
see the sequence base pairs themselves coloured on the right-hand
side.</p>
<div class="alert alert-info">
<p><strong>Side note:</strong></p>
<p>We could also “Translate” our sequence into amino acids (proteins) if
we click on the “Translated Protein Sequence” button. This will prompt
us if the genetic code is “Standard” (i.e., if it is for nuclear data).
<em>It is not</em>, and so we would click “Cancel” and select
“Vertebrate Mitochondrial” because this gene is found in the
mitochondria of vertebrates!</p>
<p><img src="Media_MEGA/Translate.png"></p>
</div>
</div>
<div class="section level3">
<h3 id="align-your-data">Align your data<a class="anchor" aria-label="anchor" href="#align-your-data"></a>
</h3>
<p>Okay, now that we have our sequences in, let’s align them.
<strong>MEGA</strong> offers two common alignment protocols,
<em>ClustalW</em> and <em>MUSCLE</em>. We don’t need to go into the
details of either (and frankly I don’t know which is better, I tend to
play around with both and then manually align as needed). So, let’s just
use ClustalW.</p>
<p>Up on the top bar, select Alignment &gt; ClustalW. This will open a
dialog to help you chose alignment parameters, for which we will keep
the defaults and select “<strong>OK</strong>” and allow
<strong>MEGA</strong> to run its alignment. It may look like your
alignment has not worked (see third screenshot below), however, if you
scroll to the right you will find that one sequence is just much longer
than the others (just as in my <strong>R</strong> example from prac
3)!</p>
<ol style="list-style-type: decimal">
<li><img src="Media_MEGA/align1.png"></li>
<li><img src="Media_MEGA/align2.png"></li>
<li><img src="Media_MEGA/align3.png"></li>
</ol>
<p>One strength of a GUI interface like <strong>MEGA</strong> or
<strong>Geneious</strong> is that you can add or remove basepairs or
spaces to help improve your alignment. This may sound like cheating with
the data, but sometimes our alignment algorithms do a poor job. Also,
sometimes when you get into science you might find that, actually, some
of our decisions are straight-up arbitrary! Not a comforting thought,
but you should always try to be the best scientist that you can be
regardless!</p>
<p>Anywho. At this point, you could export your alignment as a .fasta
file and then load it back into R, if you would like (as below). Or, you
could continue on to making a tree in <strong>MEGA</strong></p>
<p><img src="Media_MEGA/align4.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="tree-building-in-mega">Tree-building in <strong>MEGA</strong><a class="anchor" aria-label="anchor" href="#tree-building-in-mega"></a>
</h2>
<div class="section level3">
<h3 id="build-your-tree">Build your tree<a class="anchor" aria-label="anchor" href="#build-your-tree"></a>
</h3>
<p>In order to build a tree in <strong>MEGA</strong>, we need to send
our new alignment back to the main session by clicking on the little
save icon in the top bar.</p>
<p><img src="Media_MEGA/saveicon.png"></p>
<p><img src="Media_MEGA/saveSession.png"></p>
<p>We can then go back to the main session and open the .mas file that
we just saved again using the “<strong>DATA</strong>” button, as above,
but this time select “Analyze”. Our data will appear in the main window
and we can then click on the “<strong>PHYLOGENY</strong>” button to make
a tree with the data that we just imported! We will use the simple
neighbour-joining tree in this example (select “Ok” to use the currently
active data). At this point we will make some decisions about how to
build the tree. We will:</p>
<ol style="list-style-type: decimal">
<li>Change the “<strong>Test of Phylogeny</strong>” to the “Bootstrap
method” with 500 replicates — this will sub-sample the data 500 times,
re-make the tree 500 times, and see how many times the topology remains
constant!</li>
<li>Change the “<strong>Substitutions Type</strong>” to a simple
“Jukes-Cantor model”.</li>
</ol>
<p>Otherwise, we will leave the defaults and click “Ok”.
<strong>Note:</strong> the “<strong>Select Codon Positions</strong>”
option may only be present if you have chosen a protein translation.</p>
<ol style="list-style-type: decimal">
<li><img src="Media_MEGA/opensession1.png"></li>
<li><img src="Media_MEGA/Phylogeny1.png"></li>
</ol>
<p>This will open a new window with two tabs; the “Original tree” and
the “Bootstrap Consensus Tree”. We want to select the latter, which will
show the below tree, with the bootstrap support percentage for how often
each node remained the same when sub-sampled. You will also note that
<strong>MEGA</strong> has kindly reported how your tree was made, which
would be critical to report in a paper… or assignment!</p>
<ol start="3" style="list-style-type: decimal">
<li><img src="Media_MEGA/phylogeny2.png"></li>
</ol>
<p>Take a moment to play with the settings here on the top bar, the
left-hand bar, and even clicking on the tree and the nodes (or
right-clicking) to see options!</p>
</div>
<div class="section level3">
<h3 id="save-your-tree">Save your tree<a class="anchor" aria-label="anchor" href="#save-your-tree"></a>
</h3>
<p>Now it’s time to save your tree and to conduct you analysis back in
<strong>R</strong>! Select File &gt; “Export Current Tree (Newick)” and
save the tree how you’d like it saved, including Bootstrap values!
<strong>MEGA</strong> will open a new window with a text string (this is
the newick format for a tree). You can go ahead and use the save icon to
chose where you want the file to be saved (or you could even straight-up
copy and paste it into a text file or probably even
<strong>R</strong>).</p>
<ol start="3" style="list-style-type: decimal">
<li><img src="Media_MEGA/savePhylogeny.png"></li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="read-your-tree-back-into-r">Read your tree back into <strong>R</strong><a class="anchor" aria-label="anchor" href="#read-your-tree-back-into-r"></a>
</h2>
<p>It’s time to re-join the world of the coding! To do this, we will
need to read our newick tree back into <strong>R</strong> and into a
“phylo” format for use with the <strong>phytools</strong> package. This
is very easily achieved using the function
<code><a href="https://rdrr.io/pkg/phytools/man/read.newick.html" class="external-link">phytools::read.newick()</a></code>!</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># read in the newick tree from the file </span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">phytools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phytools/man/read.newick.html" class="external-link">read.newick</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"MEGA_tree.nwk"</span><span class="op">)</span></span>
<span>  <span class="co"># Read in the neqick tree from text! </span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">phytools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phytools/man/read.newick.html" class="external-link">read.newick</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">"(((((((Asemospiza_obscura_HQ153059,Geospiza_fortis_21616577)0.8160,Geospiza_difficilis_696154205)0.5400,(Certhidea_fusca_696154149,Certhidea_olivacea_696154217)0.5640)0.9300,Geospiza_septentrionalis_696154211)0.8000,Geospiza_magnirostris_696154197,Geospiza_conirostris_696154153,Geospiza_fuliginosa_696154219)0.5380,Geospiza_scandens_696154203)0.9700,Camarhynchus_heliobates_696154169,Camarhynchus_psittacula_696154121,Camarhynchus_pallidus_696154183Camarhynchus_parvulus_696154177,Camarhynchus_pauper_696154127);"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Plot the tree using phytools</span></span>
<span><span class="fu">phytools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phytools/man/plotTree.html" class="external-link">plotTree</a></span><span class="op">(</span><span class="va">tree</span>, ftype <span class="op">=</span> <span class="st">"i"</span>, fsize <span class="op">=</span> <span class="fl">0.5</span>, lwd <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="co"># Add in the nodes</span></span>
<span><span class="fu">phytools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phytools/man/labelnodes.html" class="external-link">labelnodes</a></span><span class="op">(</span>text <span class="op">=</span> <span class="va">tree</span><span class="op">$</span><span class="va">node.label</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>                     node <span class="op">=</span> <span class="va">tree</span><span class="op">$</span><span class="va">edge</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,</span>
<span>                     interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                     shape <span class="op">=</span> <span class="st">"circle"</span>,</span>
<span>                     cex <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="MEGA_tutorial_files/figure-html/readInNewick-1.png" width="672"></p>
<p>That will just about catch you up to where we are in practical 4!</p>
<div class="alert alert-danger" role="alert">
<p><strong>Assignment note:</strong></p>
<p>Make certain that you have all of the information that you would need
to report how you did this in your assignment.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James B. Dorey.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
