<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-gb">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Evolution Prac 2 — Access genetic data from GenBank • Evolution Practicals</title>
<!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Evolution Prac 2 — Access genetic data from GenBank">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Evolution Practicals</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html" aria-label="Go to package home"><span class="fa fa-home"></span></a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Available pages"><span class="fa fas fa-book"></span> Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/PhyloPrac1.html">Evolution Prac 1 — Getting ready!</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac2.html">Evolution Prac 2 — Access genetic data from GenBank</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac3.html">Evolution Prac 3 — Aligning genetic data and building phylogenies in R</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac4.html">Evolution prac 4 — Macroevolutionary questions!</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac5.html">Evolution prac 5 — Tips and pitfalls</a></li>
    <li><a class="dropdown-item" href="../articles/MEGA_tutorial.html">Evolution Prac MEGA supplement — Using MEGA to align sequences and build trees</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">


<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Evolution Prac 2 — Access genetic data from GenBank</h1>
                        <h4 data-toc-skip class="author">Dr James B
Dorey and Dr Damien Esquerré, University of Wollongong</h4>
            
            <h4 data-toc-skip class="date">Version 2025-07-18; practical
2025-08-08</h4>
      

      <div class="d-none name"><code>PhyloPrac2.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      R is a very useful programming language that’s open access and
      widely used in science. We don’t expect you to become a master by
      the end of this topic or to understand everything that you’re
      doing right now. But, by helping you get started and giving you
      robust example code we hope to enable you to succeed using R in
      the future, should you chose to pick it up more consistently.
      Here, we will use it as an <em>example</em> of how you can run
      genetic/phylogenetic analyses on a single platform. But note,
      others do exist.<br><strong>Be certain to save your R script from this week to help
      you next week!</strong>
    </div>
    
<div class="alert alert-info">
<p><strong> Attention:</strong> <br><strong>Once you have completed all
of the questions</strong> and <strong>shown your demonstrator or
Lecturer</strong> the answers and the outputs, you may leave. Or, feel
free to hang around, play, or help your classmates.</p>
</div>
<div class="section level2" number="1">
<h2 id="script-preparation">
<span class="header-section-number">1</span> Script preparation<a class="anchor" aria-label="anchor" href="#script-preparation"></a>
</h2>
<div class="section level3" number="1.1">
<h3 id="working-directory">
<span class="header-section-number">1.1</span> Working
directory<a class="anchor" aria-label="anchor" href="#working-directory"></a>
</h3>
<p>To start with, let’s set up a working directory in a place that makes
sense. We will use this working directory to, preferably, keep ALL of
the data and scripts associated with this project. This is because, if
we were to start putting files and scripts all over the computer, we
increase the chances of losing them as we change file names and paths
and, if we want to share the whole project, we can share everything by
compressing and sharing one folder. <strong>Save this folder for next
week!</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Set the RootPath to EPrac2 folder</span></span>
<span><span class="va">RootPath</span> <span class="op">&lt;-</span> <span class="st">"YourFolderPathHere"</span></span>
<span>  <span class="co"># You can then set this as the project's working directory. </span></span>
<span>  <span class="co"># This is where R will first look to find </span></span>
<span>  <span class="co"># or save data as a default</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="va">RootPath</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3" number="1.2">
<h3 id="install-packages">
<span class="header-section-number">1.2</span> Install packages<a class="anchor" aria-label="anchor" href="#install-packages"></a>
</h3>
<p>R is a programming language which has a <a href="https://stat.ethz.ch/R-manual/R-devel/library/base/html/00Index.html" class="external-link"><em>base</em></a>
and several default packages. But, because R is an open-source
programming language, it currently has <em>20,853</em> packages
available on <a href="https://cran.r-project.org/web/packages/" class="external-link">“CRAN”</a>; the official
R package repository. Additional reliable R packages can be sourced from
<a href="https://www.bioconductor.org/packages/release/BiocViews.html#___Software" class="external-link">bioconductor</a>
(<em>2,300</em> packages for working with biological data) and countless
others on GitHub.</p>
<p>This is all to say that the possibilities of analyses in R are nearly
endless, free, open source, and widely used. This makes R an incredibly
powerful language to learn for your education, research, and future
career.</p>
<p>To start out, we will need to install some basic packages. In
particular, we will install packages in the <em>tidyverse</em> which
have benefits over R’s <em>base</em> packages in that they tend to be
quicker and, more importantly, much easier to code with and
understand.</p>
<p>You’ll notice that we are actually defining an R object as a simple
character vector called “packageList” and then feeding it into the
<em>base</em> R <code><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages()</a></code> function.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">packageList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="co"># Tidyverse packages:</span></span>
<span>  <span class="st">"dplyr"</span>,            <span class="co">#  For data management and table manipulation</span></span>
<span>  <span class="st">"magrittr"</span>          <span class="co">#  to use tidy pipes; %&gt;%</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># List the new (not installed) packages and then if there are any, install them.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span>pkgs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">packageList</span><span class="op">)</span>, </span>
<span>              rebuild <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>              repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Media_EP2/dplyr.png" width="100"><img src="Media_EP2/magrittr.png" width="100"></p>
<p><br><br></p>
<p>While you’re here, why don’t you run just the below word in your
console and see what comes up. You can then also run that same object in
the <em>base</em> R <code><a href="https://rdrr.io/r/base/class.html" class="external-link">class()</a></code> function to see what type it
is.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">packageList</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">packageList</span><span class="op">)</span></span></code></pre></div>
<div class="alert alert-warning" role="alert">
<blockquote>
<p><strong>Q1:</strong> How does your input for “packageList” vary from
what comes out when you run it in the R console?</p>
</blockquote>
</div>
<div class="alert alert-warning" role="alert">
<blockquote>
<p><strong>Q2:</strong> What is the class of “packageList”?</p>
</blockquote>
</div>
</div>
<div class="section level3" number="1.3">
<h3 id="load-packages">
<span class="header-section-number">1.3</span> Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<p>Even though we have installed the packages, we actually need to
explicitly call each package using the <code><a href="https://rdrr.io/r/base/library.html" class="external-link">library()</a></code> function.
This tells R which packages it should use and can solve issues of one
function being called by multiple packages.</p>
<p>So, let’s go ahead and load in the packages that we just
installed.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Load all packages from the list specified above</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">packageList</span><span class="op">)</span>, </span>
<span>       <span class="va">library</span>, character.only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>I prefer to go one step further, which removes the issue almost
entirely, and is useful when going over your scripts and making it easy
to find errors and understand what’s happening. You can, at each
function, explicitly call the package using two colons —
<code>package**::**function()</code>. For example:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># We can explicitly call the base package's "library" function.</span></span>
<span>  <span class="co"># However, I tend to not do this with the base and default packages.</span></span>
<span><span class="fu">base</span><span class="fu">::</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span></code></pre></div>
<p>Now, that’s probably enough of an introduction to R. We can work
through more of that at a later time and in the course of this topic.
Now, let’s get onto the subject at hand… Harvesting genetic data.</p>
</div>
</div>
<div class="section level2" number="2">
<h2 id="accessing-genbank-over-a-browser">
<span class="header-section-number">2</span> Accessing GenBank over
a browser<a class="anchor" aria-label="anchor" href="#accessing-genbank-over-a-browser"></a>
</h2>
<p>There are several places online to download published genetic data,
but for life sciences, <a href="https://www.ncbi.nlm.nih.gov/genbank/" class="external-link">GenBank</a> is probably the
major host. But, it is not the sole repository. Especially for the
increasing amounts of next generation sequencing datasets, authors are
hosting data on generalist repositories like <a href="https://figshare.com" class="external-link">FigShare</a>.</p>
<p>We are <em>also</em> going to access genetic data from GenBank using
R. However, I think its still very much worthwhile taking an initial
look on the website itself and this is how I would personally access
these data in the first instance. Navigate to the <a href="https://www.ncbi.nlm.nih.gov" class="external-link">GenBank</a> homepage now. At the top
of the page, there is a search bar, there search “<strong>All
Databases</strong>” for one of Darwin’s finches, “<em>Geospiza
conirostris</em>” (<strong>Fig. 1</strong>).</p>
<div class="float">
<img src="Media_EP2/search.png" alt="The GenBank search bar."><div class="figcaption">The GenBank search bar.</div>
</div>
<p>This search should return something like the below (<strong>Fig.
2</strong>). Now, there’s quite a lot to see there and GenBank has
broken it down into the literature sources and citations (which is very
helpful!). It also provides data in the form of genomes from various
sources, genes, genetic data associated with health issues
(“<em>Clinical</em>”), and proteins. Basically, GenBank is seeking to be
able to host and share a whole bunch of molecular data with anyone who
wants to use it. Great open science! <strong>Just remember, when you use
these data you must cite the original source!</strong> Good thing that
GenBank gives the citation to you.</p>
<div class="section level3" number="2.1">
<h3 id="searching-a-taxon">
<span class="header-section-number">2.1</span> Searching a
taxon<a class="anchor" aria-label="anchor" href="#searching-a-taxon"></a>
</h3>
<p>Under <strong>Genes</strong>, let’s click on <strong>PopSet</strong>.
Here you will see a whole bunch of single gene region extracts from
finches of this species, information about the gene that has been
sequenced (e.g., MHC class IIB antigen pseudogene, cytochrome b,
ornithine decarboxylase, and other stupid gene names…), and information
about those extracts. Underneath there is more information to help you
decide what you might like to look at, such as the number of sequences
provided and the type of study.</p>
<div class="float">
<img src="Media_EP2/Screenshot%202024-06-06%20at%2010.03.05%20am.png" alt="The search results from GenBank for the finch, “Geospiza conirostris”."><div class="figcaption">The search results from GenBank for the finch,
“<em>Geospiza conirostris</em>”.</div>
</div>
<p>Let us click on “Passeriformes MHC class IIB antigen pseudogene,
partial sequence.” and look a little closer.</p>
<div class="alert alert-warning" role="alert">
<blockquote>
<p><strong>Q3:</strong> How many sequences are in this alignment?</p>
</blockquote>
</div>
<div class="alert alert-warning" role="alert">
<blockquote>
<p><strong>Q4:</strong> What is the <em>title</em> and <em>journal</em>
of the study that produced these genes and shared them on GenBank?</p>
</blockquote>
</div>
<div class="alert alert-warning" role="alert">
<blockquote>
<p><strong>Q5:</strong> For the gene with the accession number
“<em>AY064447.1</em>”, click on that number and tell us, what are the
first 10 basepairs for the sequence? <em>HINT: have a look down the
bottom of the page.</em></p>
</blockquote>
</div>
<div class="alert alert-warning" role="alert">
<blockquote>
<p><strong>Q6:</strong> On the alignment page, can you find and download
the entire alignment with all sequnces in a <strong>FASTA</strong>
format? <em>HINT: search the page for “FASTA” and click on that link; on
that page at the top right use the “Send to:” drop down to select “File”
and then click “Create File”.</em></p>
</blockquote>
</div>
<div class="alert alert-warning" role="alert">
<blockquote>
<p><strong>Q7:</strong> What does the fasta file look like and what
information does it contain?</p>
</blockquote>
</div>
<p>Fasta files are one of the most commonly used formats for genetic
data. They are nice and simple files that are essentially just plain
text that store two data types (<strong>Q7</strong>). Their simplicity
makes them powerful and easy to understand.</p>
</div>
<div class="section level3" number="2.2">
<h3 id="searching-a-group">
<span class="header-section-number">2.2</span> Searching a
group<a class="anchor" aria-label="anchor" href="#searching-a-group"></a>
</h3>
<p>Now, let’s make a more useful search. Let us try and find the data
that’s available for all of Darwin’s finches in the genera
<em>Geospiza</em>, <em>Cactospiza</em>, <em>Camarhynchus</em>, and
<em>Certhidea</em>.</p>
<p>Go back to the <a href="https://www.ncbi.nlm.nih.gov" class="external-link">GenbBank</a>
homepage and instead of searching “<strong>All Databases</strong>”,
search “<strong>Nucleotide</strong>” and click “<strong>Search</strong>”
with the text bar empty and, on the following page, click
“<strong>Advanced</strong>”. Here, we can start to build an advanced
query. &gt; 1. In the first field, enter “cytb” which is the
mitochondrial gene “<em>cytochrome b</em>”; a very commonly used gene
for shallow (~recent divergences) inter-species phylogenetic analyses.
Choose the drop-down as “<strong>Gene Name</strong>” and not
“<strong>All Fields</strong>”.</p>
<blockquote>
<ol start="2" style="list-style-type: decimal">
<li>In the next four fields put the four bird genera names and choose
the drop-down as “<strong>Organism</strong>”</li>
</ol>
</blockquote>
<blockquote>
<ol start="3" style="list-style-type: decimal">
<li>In the “<strong>AND</strong>” drop-downs, choose
“<strong>OR</strong>” and click “<strong>Search</strong>” (<strong>Fig.
3</strong>)</li>
</ol>
</blockquote>
<div class="float">
<img src="Media_EP2/advancedSearch.png" alt="A GenBank advanced search. Note, that as you enter names into the Builder, a text version of your search query is built directly above!"><div class="figcaption">A GenBank advanced search. Note, that as you
enter names into the Builder, a text version of your search query is
built directly above!</div>
</div>
<blockquote>
<ol start="4" style="list-style-type: decimal">
<li>Because we have chosen all “<strong>OR</strong>”s, we actually have
cytB genes for any species at all! Oops! However, we can fix this and
make a much better search with this In the search bar at the top, you
can see that our search has been turned into a text query!</li>
</ol>
</blockquote>
<blockquote>
<p>“((((cytb[Gene Name]) OR Geospiza[Organism]) OR Cactospiza[Organism])
OR Camarhynchus[Organism]) OR Certhidea[Organism]”</p>
</blockquote>
<p>Here you can see that our query is bounded by a bunch of confusing
brackets. They are specifying the relationships between each query, but
what we want is to find all cytB genes for any (<strong>OR</strong>) of
the four genera, so we can enclose the gene name and then all of the
genera, separated by an “<strong>AND</strong>” to specify this. Go ahead
and make this change or paste the below query into the search bar.</p>
<blockquote>
<p>“<strong>(</strong>cytb[Gene Name]<strong>)</strong>
<strong>AND</strong> <strong>(</strong>”Geospiza”[Organism]) OR
“Camarhynchus”[Organism] OR “Certhidea”[Organism]<strong>)</strong>”</p>
</blockquote>
<p>Now, you’ll see that there are some crazy genetic materials in there
including genomic data (genes from in the cell nucleus). On the
left-hand side there’s a filter that you can choose called
“<strong>Mitochondrion</strong>” (search for it on the page if you
need). Click on that and you’ll see that what’s mostly left are a bunch
of 800 to 1,120 basepair sequences. But, there are also some much long
full mitogenomes (&gt;16,000 bp) and sequences of different lengths
(probably differing due to the sequencing and filtering methods used in
their production). You may also note that “<em>AND
mitochondrion[filter]</em>” was also added to the search query.</p>
</div>
<div class="section level3" number="2.3">
<h3 id="download-sequence-data-from-the-browser">
<span class="header-section-number">2.3</span> Download sequence
data from the browser<a class="anchor" aria-label="anchor" href="#download-sequence-data-from-the-browser"></a>
</h3>
<p>From this point, you might go through these sequences and try to find
one for each species you want to include, or a subset of those species.
You might sort these sequences by length (<strong>Fig. 4</strong>) and
then look for the longest sequence of each species as you scroll down
the page.</p>
<div class="float">
<img src="Media_EP2/sortLength.png" alt="Sort GenBank results by length also note that’s i’ve chosen to display 200 records per page."><div class="figcaption">Sort GenBank results by length also note that’s
i’ve chosen to display 200 records per page.</div>
</div>
<p>Why don’t you take a few moments to try and find the longest
sequences for all of the species that have cytB data on GenBank?
<strong>Hint</strong>: there are 14 species that I can see cytB for:
<em>Geospiza magnirostris</em>, <em>Geospiza fortis</em>,
<em>Camarhynchus heliobates</em>, <em>Geospiza difficilis</em>,
<em>Geospiza conirostris</em>, <em>Certhidea fusca</em>, <em>Certhidea
olivacea</em>, <em>Camarhynchus psittacula</em>, <em>Camarhynchus
parvulus</em>, <em>Camarhynchus pauper</em>, <em>Geospiza
fuliginosa</em>, <em>Geospiza scandens</em>, <em>Geospiza
septentrionalis</em>, and <em>Camarhynchus pallidus</em>.</p>
<p>You could download each as a fasta file and then copy them together
in a text editor, or just click on “FASTA” for each sequence and copy
the fasta text into a text file (<strong>Fig. 5</strong>).
<strong>Note</strong>: for mac users <a href="https://apps.apple.com/us/app/bbedit/id404009241?mt=12" class="external-link"><strong>BBEdit</strong></a>
is a fantastic and free text editor (<strong>Fig. 6</strong>).</p>
<div class="float">
<img src="Media_EP2/fastaCopy.png" alt="Copy the two lines of each fasta and paste it into a text editor (next figure)."><div class="figcaption">Copy the two lines of each fasta and paste it
into a text editor (next figure).</div>
</div>
<div class="float">
<img src="Media_EP2/bbeditFastaExample.png" alt="Here are three sequences copied into one fasta file. the lines of DNA can either be hard-wrapped (line breaks in-line; top sequence) or not (i.e., one long bit of sequence without any line breaks; bottom sequence)."><div class="figcaption">Here are three sequences copied into one fasta
file. the lines of DNA can either be hard-wrapped (line breaks in-line;
<strong>top sequence</strong>) or not (i.e., one long bit of sequence
without any line breaks; <strong>bottom sequence</strong>).</div>
</div>
<p>Save this file with an informative name in your R working directory
and with the file suffix “.fasta”. Maybe something like
<em>darwinFinches_cytB.fasta</em> — we will skip this part right now in
favor of showing you, very quickly, how to accomplish this in R.
However, reading the fasta into R is really simple using the
<strong>ape</strong> package (see the example line of code below and the
next section).</p>
<pre><code><span><span class="va">DarwinFinches_cytB</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.dna.html" class="external-link">read.FASTA</a></span><span class="op">(</span><span class="st">"darwinFinches_cytB.fasta"</span><span class="op">)</span></span></code></pre>
<p><strong>Warning:</strong> Sometimes your computer will try to save
this as <em>darwinFinches_cytB.fasta<strong>.txt</strong></em>. You
either need to remove that or include it when you read into ape.</p>
</div>
</div>
<div class="section level2" number="3">
<h2 id="accessing-genbank-in-r">
<span class="header-section-number">3</span> Accessing GenBank in
R<a class="anchor" aria-label="anchor" href="#accessing-genbank-in-r"></a>
</h2>
<div class="section level3" number="3.1">
<h3 id="examining-genbank-databases">
<span class="header-section-number">3.1</span> Examining GenBank
databases<a class="anchor" aria-label="anchor" href="#examining-genbank-databases"></a>
</h3>
<p>Now, that we’ve had an explore of GenBank on a nice browser, we can
have a look at accessing it right here in R. <em>Gasp!</em> For that,
we’re going to want some extra packages because the creator of R — being
as short-sighted as they are — did not consider us phylogeneticists when
making their programming language … &gt;:(</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Let's install some packages the old-fashioned way just to show you how</span></span>
<span><span class="co"># Accesses GenBank's "Entrez Utilities" API through R (i.e., it searches GenBank)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"rentrez"</span>,</span>
<span>              repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span>
<span><span class="co"># A central R phylogenetics package that stands for "Analysis of Phylogenetics and Evolution"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ape"</span>,</span>
<span>              repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span> </span>
<span></span>
<span>  <span class="co"># Let's do the same by loading them in</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/rentrez" class="external-link">rentrez</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/emmanuelparadis/ape" class="external-link">ape</a></span><span class="op">)</span></span></code></pre></div>
<p><img src="Media_EP2/rentrez.png" width="100"><img src="Media_EP2/ape.png" width="100"></p>
<p><br></p>
</div>
<div class="section level3" number="3.2">
<h3 id="run-a-genbank-query">
<span class="header-section-number">3.2</span> Run a GenBank
query<a class="anchor" aria-label="anchor" href="#run-a-genbank-query"></a>
</h3>
<p>Let us dive right into querying GenBank in R. First, lets use
<code><a href="https://docs.ropensci.org/rentrez/reference/entrez_search.html" class="external-link">entrez_search()</a></code> on the “nucleotide” database using our
previous search term. <strong>Note:</strong> that we have to remove the
internal quotes in the search “<strong>term</strong>”. Use the following
term: <em>“(cytb[Gene Name]) AND (Geospiza[Organism]) OR
Camarhynchus[Organism] OR Certhidea[Organism]) AND
(mitochondrion[filter])”</em></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myFirstSearch</span> <span class="op">&lt;-</span> <span class="fu">rentrez</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/rentrez/reference/entrez_search.html" class="external-link">entrez_search</a></span><span class="op">(</span>db <span class="op">=</span> <span class="st">"nucleotide"</span>,</span>
<span>              term <span class="op">=</span> <span class="st">"(cytb[Gene Name]) AND (Geospiza[Organism]) OR Camarhynchus[Organism] </span></span>
<span><span class="st">              OR Certhidea[Organism]) AND (mitochondrion[filter])"</span>,</span>
<span>                <span class="co"># With this new search term, I can see 171 sequences, which is enough for me to</span></span>
<span>                <span class="co"># download, I think! So, I'll set this in retmax</span></span>
<span>              retmax <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspect the object and answer the question below</span></span>
<span><span class="va">myFirstSearch</span></span></code></pre></div>
<div class="alert alert-warning" role="alert">
<blockquote>
<p><strong>Q8:</strong> How many search results were found and does that
number seem familiar?</p>
</blockquote>
</div>
<div class="section level4" number="3.2.1">
<h4 id="accessing-r-help-files">
<span class="header-section-number">3.2.1</span> Accessing R help
files<a class="anchor" aria-label="anchor" href="#accessing-r-help-files"></a>
</h4>
<p>Let’s take a quick interlude at this point.</p>
<p>Let us say that we are novices and we don’t know what a function can
or should do? Well, R is here to help and this is a tip that you will be
using for years to come. If you want to quickly bring up the function’s
help file simply run “?<em>function</em>”. In this case, let’s have a
closer look at <code><a href="https://docs.ropensci.org/rentrez/reference/entrez_search.html" class="external-link">entrez_search()</a></code> by typing
“?<em>entrez_search</em>” into the R console.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Access the help file for the function entrez_search</span></span>
<span><span class="op">?</span><span class="va">entrez_search</span></span></code></pre></div>
<p>Here you will see with a bunch of information to hopefully help you
understand and use the function!</p>
<div class="alert alert-warning" role="alert">
<blockquote>
<p><strong>Q9:</strong> What does the help file say under the
“<strong>Description</strong>”?</p>
</blockquote>
</div>
</div>
</div>
<div class="section level3" number="3.3">
<h3 id="download-genetic-data">
<span class="header-section-number">3.3</span> Download genetic
data<a class="anchor" aria-label="anchor" href="#download-genetic-data"></a>
</h3>
<p>We can then use the <em>ape</em> package and these accession numbers
to download the fasta data right onto our computers. We will use the
<code><a href="https://rdrr.io/pkg/ape/man/read.GenBank.html" class="external-link">read.GenBank()</a></code> function from <em>ape</em>, so why don’t you
first run “?<em>read.GenBank</em>” to have a look at the help file,
especially focus on the <strong>Description</strong>,
<strong>Arguments</strong>, and <strong>Value</strong> sections.</p>
<p><strong>Note:</strong> Most R objects are some kind of “list” and you
can access elements of that list by R operator “$”. For example, if you
type in “myFirstSearch$” in the console and wait a moment, it will come
up with a drop down list of the objects inside that object! This is
really important, because to now download the genetic data, we need the
GenBank accession numbers which are stored in the “$ids” slot of our
object.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download the data from GenBank using the ids that you got from rentrez</span></span>
<span><span class="va">DarwinFinches_cytB</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.GenBank.html" class="external-link">read.GenBank</a></span><span class="op">(</span>access.nb <span class="op">=</span> <span class="va">myFirstSearch</span><span class="op">$</span><span class="va">ids</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># What type of file is this? </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># What does it look like?</span></span>
<span><span class="va">DarwinFinches_cytB</span></span>
<span></span>
<span><span class="co"># Have a look at one sequence with the accession number "56682244"</span></span>
<span><span class="va">DarwinFinches_cytB</span><span class="op">$</span><span class="va">`56682244`</span></span>
<span>  <span class="co"># Wow, those data are stored really strangely! I agree. DNAbin objects store genetics data</span></span>
<span>  <span class="co"># in non-user-friendly ways. Many genetics programs do, in fact. I bet that you miss that</span></span>
<span>  <span class="co"># simple fasta now </span></span>
<span><span class="co"># However, this format also allows extra attributes to be saved! So, let's save this as an R</span></span>
<span>  <span class="co"># object for next week</span></span>
<span><span class="fu">base</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB</span>, </span>
<span>        file <span class="op">=</span> <span class="st">"DarwinFinches_cytB.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Let's save this fasta file</span></span>
<span><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/write.dna.html" class="external-link">write.FASTA</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB</span>,</span>
<span>                 file <span class="op">=</span> <span class="st">"DarwinFinches_cytB.fasta"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2" number="4">
<h2 id="start-thinking-about-your-project">
<span class="header-section-number">4</span> Start thinking about
your project<a class="anchor" aria-label="anchor" href="#start-thinking-about-your-project"></a>
</h2>
<div class="section level3" number="4.1">
<h3 id="think-about-a-taxon-thats-interesting-to-you">
<span class="header-section-number">4.1</span> Think about a taxon
that’s interesting to you<a class="anchor" aria-label="anchor" href="#think-about-a-taxon-thats-interesting-to-you"></a>
</h3>
<p>Now then, it’s time for you to choose a taxon that you might be
interested in looking at! You might already have a good idea about a
taxon that you’re interested in! Maybe you’re into frogs, lizards,
birds, wasps, bees, whatever! Maybe you have no idea! Regardless, I’d
suggest choosing a group that is reasonable to deal with and not
overwhelming. Maybe somewhere between 7 and 30 species? You can always
choose a bigger group as well and then take a step backwards in regards
to how many species you include from that group (you don’t need to
include them all).</p>
<p>But, how on earth will you know this? Well, I can provide a little
guidance here and a quick way for you to start querying taxa and their
size. <strong>Keep in mind</strong> that even if a group has 10 species,
you may not be able to get genetic data for all of those 10 species —
and, you need genetic data that aligns (i.e., is from the same gene and
the same part of that gene). So, maybe shoot for a few more species and
then search for what’s available.</p>
<p>The below function, <code><a href="https://jbdorey.github.io/BeeBDC/reference/taxadbToBeeBDC.html" class="external-link">taxadbToBeeBDC()</a></code> from the
<strong>BeeBDC</strong> package will get taxon names from <a href="https://www.gbif.org" class="external-link">GBIF</a>, the Global Biodiversity
Information Facility, and return a table with the valid and old names
for each species. We will then use the <strong>dplyr</strong> package to
filter to only “<strong>accepted</strong>” names, group by
“<strong>genus</strong>” and then return a table with the count of valid
species in each genus. Please, feel free to play around at this
point!</p>
<div class="alert alert-danger" role="alert">
<p><strong>Note:</strong> For some taxa, GBIS is not the ideal choice of
taxon provider. A good alternative is the <a href="https://www.itis.gov" class="external-link">Integrated Taxonomic Information System
(ITIS)</a>. You can change the below from <code>provider = "gbif"</code>
to <code>provider = "itis"</code>. Simple.</p>
</div>
<p><strong>Note:</strong> You may get an error in the below code. Please
do <em>read the error</em>, then <em>think about it</em>, and then see
if you can solve it! A big part of programming is being able to read and
solve errors. You can also totally google it or ask ChatGPT (<strong>use
ChatGPT with extreme caution; it can and will lead to more
problems</strong>)! If these fail, ask your demonstrator.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># A useful package for taxonomic and occurrence data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BeeBDC"</span>,</span>
<span>              repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jbdorey.github.io/BeeBDC/" class="external-link">BeeBDC</a></span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Run the BeeBDC query</span></span>
<span><span class="va">speciesCountTable</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="https://jbdorey.github.io/BeeBDC/reference/taxadbToBeeBDC.html" class="external-link">taxadbToBeeBDC</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Apidae"</span>,</span>
<span>                               rank <span class="op">=</span> <span class="st">"Family"</span>,</span>
<span>                               provider <span class="op">=</span> <span class="st">"gbif"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">taxonomic_status</span> <span class="op">==</span> <span class="st">"accepted"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">genus</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#As usual, what class is returned?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">speciesCountTable</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># What did we make?</span></span>
<span><span class="va">speciesCountTable</span></span>
<span></span>
<span><span class="co"># Did you know that you could sort the table? You could either look at the function:</span></span>
<span><span class="op">?</span><span class="fu">dplyr</span><span class="fu">::</span><span class="va"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span></span>
<span>  <span class="co"># or run:</span></span>
<span><span class="co">#View(speciesCountTable) # REMEMBER: R is case-sensitive!</span></span>
<span>  <span class="co"># and sort by the "n" column</span></span></code></pre></div>
<p><img src="Media_EP2/BeeBDC.png" width="100"></p>
<p><br></p>
<div class="alert alert-warning" role="alert">
<blockquote>
<p><strong>Q10:</strong> Which bee genus in the Apidae family has the
most species in it, according to GBIF?</p>
</blockquote>
</div>
<p><strong>Note:</strong> But hey, what a minute, James, what the shit
is this “%&gt;%” crap? Well, I’m glad that you <em>[were forced]</em> to
ask! It’s called a <strong>pipe</strong> and it’s from the
<em>magrittr</em> package. It essentially passes the output of the first
function forwards (<strong>%&gt;%</strong>) into the next function. It
is equivalent to running the below, but it’s much quicker and
neater.</p>
<pre><code><span>  <span class="va">BeeBDC_query</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="https://jbdorey.github.io/BeeBDC/reference/taxadbToBeeBDC.html" class="external-link">taxadbToBeeBDC</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Apidae"</span>,</span>
<span>                           rank <span class="op">=</span> <span class="st">"Family"</span>,</span>
<span>                           provider <span class="op">=</span> <span class="st">"gbif"</span><span class="op">)</span> </span>
<span>  <span class="va">BeeBDC_query</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">BeeBDC_query</span>, <span class="va">taxonomic_status</span> <span class="op">==</span> <span class="st">"accepted"</span><span class="op">)</span></span>
<span>  <span class="va">BeeBDC_query</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">BeeBDC_query</span>, <span class="va">genus</span><span class="op">)</span></span>
<span>  <span class="va">speciesCountTable</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">BeeBDC_query</span><span class="op">)</span></span>
<span>  </span></code></pre>
<p>See how in the above <code>BeeBDC_query</code> needs to get passed on
every single time into the next line and then over-written? Using pipes
can make coding much easier to read, write, and error-find!</p>
</div>
<div class="section level3" number="4.2">
<h3 id="what-genetic-markers-are-out-there">
<span class="header-section-number">4.2</span> What genetic markers
are out there?<a class="anchor" aria-label="anchor" href="#what-genetic-markers-are-out-there"></a>
</h3>
<p>Okay, I’ve given you the tools that you’ll need to choose a taxonomic
group and to query GenBank to get your data. The rest is up to you. If
you still have time in this prac, please start playing around with all
of this and searching for a group that might interest you. Additionally,
there are more genes to choose from! We used <strong>cytB</strong> here,
but there are loads of other very commonly used genes. Some examples
include:</p>
<ol style="list-style-type: decimal">
<li>
<strong>COI</strong> — Cytochrome c oxidase subunit I
(mitochondrial)</li>
<li>
<strong>EEF1A1</strong> — eukaryotic translation elongation factor 1
alpha 1 (nuclear)</li>
<li>
<strong>16S</strong> rRNA (ribosomal RNA)</li>
<li>
<strong>28S</strong> rRNA (ribosomal RNA)</li>
<li>
<strong>ITS</strong> rRNA — Internal transcribed spacer (ribosomal
RNA)</li>
</ol>
<p>There are plenty more. So, choose your taxon and have a look at
what’s available! Especially this initial search might be better done
using the online version of GenBank</p>
<div class="alert alert-danger" role="alert">
<blockquote>
<p><strong>Q12:</strong> What is a…</p>
<ul>
<li>taxon that’s interesting to you;</li>
<li>that might have an appropriate number of species;</li>
<li>that has an appropriate number of genetic sequences;</li>
<li>and what gene region could fit that criteria?</li>
</ul>
</blockquote>
</div>
</div>
</div>
<div class="section level2" number="5">
<h2 id="whats-coming-next-week">
<span class="header-section-number">5</span> What’s coming next
week?<a class="anchor" aria-label="anchor" href="#whats-coming-next-week"></a>
</h2>
<p>Next week, we will be putting the genetic data that we gather this
week together! We will begin with aligning the sequences together and
then we will begin phylogenetic inference (making phylogenetic trees).
So, do try and make sure you’ve got something for next week. Even if
it’s for just Darwin’s finches.</p>
<p><strong>Be certain to save your R script from this week to help you
next week!</strong> If you don’t, well… sounds like you’ve got some
catching up to do.</p>
<div class="alert alert-warning" role="alert">
<p><strong>Attention:</strong> By the end of Week 4, I want you to have
the data that you mentioned in Question 12. You will need to show this
to your demonstrator.</p>
</div>
<p><br><br><br><br></p>
</div>
<div class="section level2" number="6">
<h2 id="packages-used-today">
<span class="header-section-number">6</span> Packages used
today<a class="anchor" aria-label="anchor" href="#packages-used-today"></a>
</h2>
<p><img src="Media_EP2/dplyr.png" width="200"><img src="Media_EP2/magrittr.png" width="200"><img src="Media_EP2/rentrez.png" width="200"><img src="Media_EP2/ape.png" width="200"><img src="Media_EP2/BeeBDC.png" width="200"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James B. Dorey.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
