<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-gb">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Evolution prac 5 — Tips and pitfalls • Evolution Practicals</title>
<!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Evolution prac 5 — Tips and pitfalls">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Evolution Practicals</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html" aria-label="Go to package home"><span class="fa fa-home"></span></a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Available pages"><span class="fa fas fa-book"></span> Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/PhyloPrac1.html">Evolution Prac 1 — Getting ready!</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac2.html">Evolution Prac 2 — Access genetic data from GenBank</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac3.html">Evolution Prac 3 — Aligning genetic data and building phylogenies in R</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac4.html">Evolution prac 4 — Macroevolutionary questions!</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac5.html">Evolution prac 5 — Tips and pitfalls</a></li>
    <li><a class="dropdown-item" href="../articles/MEGA_tutorial.html">Evolution Prac MEGA supplement — Using MEGA to align sequences and build trees</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">


<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Evolution prac 5 — Tips and pitfalls</h1>
                        <h4 data-toc-skip class="author">Dr James B
Dorey and Dr Damien Esquerré, University of Wollongong</h4>
            
            <h4 data-toc-skip class="date">Version 2025-07-18; practical
2025-08-29</h4>
      

      <div class="d-none name"><code>PhyloPrac5.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      This prac manual has been put together in a sorta rushed manner
      just to give you all a little push in the right direction for some
      common pitfalls and to try and get you on your way with your own
      assignment. Hopefully it is helpful, it is in no way exhaustive.
      Please read it through or search it to help you along.
      <strong>Here’s the thing as well, folks, one very important skill
      in coding is being able to FIND YOUR OWN SOLUTIONS! Google yout
      error first and you might get a good answer right away!</strong>
      One of R’s strenghts, if you’ll remember, is that there are MANY
      people using it and so many people have come before and probably
      experienced the same problem that you have. Don’t be afraid to
      have a little look around :)<br><strong>Also, MAKE SURE that you have read the assignment
      guidelines. I wrote these to help guide you and gave yo ua rubric
      so you know EXACTLY how you can get full marks which, by the way,
      I would be <em>very</em> happy to give!</strong>
    </div>
    
<div class="section level2" number="1">
<h2 id="script-preparation">
<span class="header-section-number">1</span> Script preparation<a class="anchor" aria-label="anchor" href="#script-preparation"></a>
</h2>
<div class="section level3" number="1.1">
<h3 id="working-directory">
<span class="header-section-number">1.1</span> Working
directory<a class="anchor" aria-label="anchor" href="#working-directory"></a>
</h3>
<p>Below is the example path to MY ROOT directoy. <strong>NOTE</strong>:
A common pitfall that I have seen is people including a <em>file</em> at
the end of the path, such as:
“<em>/Users/jamesdorey/Desktop/Uni/Teaching/BIOL365_971/BIOL365_pracs_2025/EPrac5<strong>/Prac5_BIOL365.Rmd</strong></em>”
See how that last part in bold is actually a file with a file extension?
That’s not a directory (folder).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Set the RootPath to Prac1 folder</span></span>
<span>  <span class="co"># This is the example path to MY FOLDER. </span></span>
<span><span class="va">RootPath</span> <span class="op">&lt;-</span> <span class="st">"/Users/jamesdorey/Desktop/Uni/Teaching/BIOL365_971/BIOL365_pracs_2025/EPrac5"</span></span>
<span>  <span class="co"># You can then set this as the project's working directory. </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="va">RootPath</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3" number="1.2">
<h3 id="install-packages">
<span class="header-section-number">1.2</span> Install packages<a class="anchor" aria-label="anchor" href="#install-packages"></a>
</h3>
<p>Below are ALL OF THE PACKAGES that we have used in these pracs</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CRAN_packageList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="co"># Tidyverse packages:</span></span>
<span>  <span class="st">"dplyr"</span>,</span>
<span>  <span class="st">"magrittr"</span>, <span class="co"># This package is critical if R says that it can't find the function "%&gt;%"</span></span>
<span>  <span class="st">"tibble"</span>,            <span class="co">#  A lovely table format package</span></span>
<span>  <span class="st">"tidyr"</span>,             <span class="co"># Another data maniupulation package from the tidyverse</span></span>
<span>  <span class="st">"stringr"</span>,           <span class="co">#  to manipulate text strings</span></span>
<span>  <span class="st">"readxl"</span>,  <span class="co"># Lets you read in excel files</span></span>
<span>    <span class="co"># GENETICS packages on CRAN</span></span>
<span>  <span class="st">"rentrez"</span>,</span>
<span>  <span class="st">"ape"</span>,</span>
<span>  <span class="st">"msaR"</span>,</span>
<span>  <span class="st">"phangorn"</span>,</span>
<span>  <span class="st">"phytools"</span>,</span>
<span>  <span class="st">"corHMM"</span>,</span>
<span>    <span class="co"># General R package to help with phytools</span></span>
<span>  <span class="st">"plotrix"</span>,</span>
<span>    <span class="co"># General R package to help with BeeBDC</span></span>
<span>  <span class="st">"taxadb"</span>,</span>
<span>    <span class="co"># This package gives you access to "bioconductor" packages</span></span>
<span>  <span class="st">"BiocManager"</span></span>
<span><span class="op">)</span></span>
<span>  <span class="co"># Install the packages that are hosted on CRAN</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span>pkgs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">CRAN_packageList</span><span class="op">)</span>, </span>
<span>              rebuild <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>              repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># These packages are installed from bioconductor</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"Biostrings"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"msa"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># READ in your packages</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">CRAN_packageList</span>, <span class="st">"dplyr"</span>, <span class="st">"magrittr"</span>, <span class="st">"ape"</span>, <span class="st">"rentrez"</span><span class="op">)</span>, </span>
<span>       <span class="va">library</span>, character.only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3" number="1.3">
<h3 id="load-packages">
<span class="header-section-number">1.3</span> Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<p>Once again, we will need to load all of our packages into R.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">CRAN_packageList</span>, <span class="st">"Biostrings"</span>, <span class="st">"msa"</span><span class="op">)</span>, </span>
<span>       <span class="va">library</span>, character.only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2" number="2">
<h2 id="do-my-genetic-data-relate-to-my-trait">
<span class="header-section-number">2</span> Do my genetic data
relate to my trait?<a class="anchor" aria-label="anchor" href="#do-my-genetic-data-relate-to-my-trait"></a>
</h2>
<p>I’ve driven this home a few times already, so have a look at the last
two prac intros for example, but… We will not be linking traits directly
to our genetic variants/mutations. This is possible but extremely
complicated and usually done in <em>genome-wide association
studies</em>. Bugger that, I say. Instead, we will:</p>
<ol style="list-style-type: decimal">
<li><p>Build a tree using the genetic data</p></li>
<li><p>Using that tree and statistical methods we will infer how our
trait(s) have changed with evolution</p></li>
</ol>
</div>
<div class="section level2" number="3">
<h2 id="what-if-i-want-to-do-a-more-complicated-analysis">
<span class="header-section-number">3</span> What if I want to do a
more complicated analysis?<a class="anchor" aria-label="anchor" href="#what-if-i-want-to-do-a-more-complicated-analysis"></a>
</h2>
<p>Sure, maybe I haven’t shown you how to do all tests and I didn’t want
to freak you out. Usually, if you want to do a standard statistical test
you can do it in a phylogenetically explicit manner as well. So,
consider the data that you have (continuous, categorical, count…), think
of (or just google) the statistical test that you will need and then
look for a function (probably in the <em>phytools</em> package).</p>
<p>Some below examples could be:</p>
<ol style="list-style-type: decimal">
<li><p>ANOVA (continuous and categorical) —
<code><a href="https://rdrr.io/pkg/phytools/man/phylANOVA.html" class="external-link">phytools::phylANOVA()</a></code></p></li>
<li><p>Chi-squared (count data — two binary traits) —
<code><a href="https://rdrr.io/pkg/phytools/man/fitPagel.html" class="external-link">phytools::fitPagel()</a></code></p></li>
</ol>
<p>Here is a link to a useful workshop in using <em>phytools</em> — <a href="http://www.phytools.org/Cordoba2017/" class="external-link uri">http://www.phytools.org/Cordoba2017/</a> — Maybe you’ll find
something cool there. But for example, <a href="http://www.phytools.org/Cordoba2017/ex/15/Plotting-methods.html" class="external-link">Exercise
15</a> has some really cool plotting ideas.</p>
</div>
<div class="section level2" number="4">
<h2 id="common-general-issues">
<span class="header-section-number">4</span> Common general
issues<a class="anchor" aria-label="anchor" href="#common-general-issues"></a>
</h2>
<div class="section level3" number="4.1">
<h3 id="troubleshooting">
<span class="header-section-number">4.1</span> Troubleshooting<a class="anchor" aria-label="anchor" href="#troubleshooting"></a>
</h3>
<p>Many issues you could figure out yourself. A very powerful technique
to consider is to select and run each R object individually and work
your way back up the code until you find the problem. For example, maybe
you made an object a few lines up, but you accidentally removed all of
the spaces from your species names in your alignment but did not
realise! Oops. When you try to make the tree with your outgroup you
might find that it says that the outgroup is not in your tree! If you
looked at the objects that got you to this point, you would probably
realise this and go to find that problem!</p>
</div>
<div class="section level3" number="4.2">
<h3 id="could-not-find-function">
<span class="header-section-number">4.2</span> could not find
function “%&gt;%”<a class="anchor" aria-label="anchor" href="#could-not-find-function"></a>
</h3>
<p>This is a common error that I get regularly. It’s also INCREDIBLY
simple to fix! The problem is that the pipe function,
<code>%&gt;%</code>, is exported from the <em>magrittr</em> package! So,
when you get the “could not find function”%&gt;%“” error, all you need
to do is run the below and all will be well again!</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3" number="4.3">
<h3 id="error-in-loadnamespacex-there-is-no-package-called-package-name">
<span class="header-section-number">4.3</span> Error in
loadNamespace(x) : there is no package called ‘PACKAGE NAME’<a class="anchor" aria-label="anchor" href="#error-in-loadnamespacex-there-is-no-package-called-package-name"></a>
</h3>
<p>Another very common issue that people came across involves R not
being able to find a particular package. For example, in the taxa name
query code below, people were often confronted with the error:</p>
<pre><code>Loading required namespace: taxadb
Failed with error:  ‘there is no package called ‘taxadb’’
 - Downloading taxonomy...
Error in loadNamespace(x) : there is no package called ‘taxadb’</code></pre>
<p>R has tried to tell you EXACTLY what the problem is here, that the
<em>taxadb</em> package is not installed! All you need to do is install
it and all will be well again. <em>For further context</em>, the reason
that this happens is because in BeeBDC, the function
<code><a href="https://jbdorey.github.io/BeeBDC/reference/taxadbToBeeBDC.html" class="external-link">BeeBDC::taxadbToBeeBDC()</a></code> is not considered to be a core
package that all users will want. To save EVERYONE from having to
download <em>taxadb</em>, we let users choose to download it if they
need it.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># A useful package for taxonomic and occurrence data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BeeBDC"</span><span class="op">)</span></span>
<span>  <span class="co"># ALSO install taxadb</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"taxadb"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Run the BeeBDC query</span></span>
<span><span class="va">speciesCountTable</span> <span class="op">&lt;-</span> <span class="fu">BeeBDC</span><span class="fu">::</span><span class="fu"><a href="https://jbdorey.github.io/BeeBDC/reference/taxadbToBeeBDC.html" class="external-link">taxadbToBeeBDC</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Apidae"</span>,</span>
<span>                               rank <span class="op">=</span> <span class="st">"Family"</span>,</span>
<span>                               provider <span class="op">=</span> <span class="st">"gbif"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">taxonomic_status</span> <span class="op">==</span> <span class="st">"accepted"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">genus</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3" number="4.4">
<h3 id="trouble-with-installing-packages">
<span class="header-section-number">4.4</span> Trouble with
installing packages<a class="anchor" aria-label="anchor" href="#trouble-with-installing-packages"></a>
</h3>
<p>Sometimes you will have trouble installing R packages or loading them
in. What a drag! Often the easiest option is to just restart R! This is
quick and easy from the “Session” drop down in R studio. See
<strong>Figure 1</strong> below.</p>
<p>An advanced option is to actually set up a new folder to save your
packages in. This is actually really easy and can be useful when you
want different places to save packages for specific research projects.
You can do the below:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Install renv = "R environment"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"renv"</span><span class="op">)</span></span>
<span>  <span class="co"># Initialise a project in your RooPath. Hey, nice.</span></span>
<span><span class="fu">renv</span><span class="fu">::</span><span class="fu">init</span><span class="op">(</span><span class="va">RootPath</span><span class="op">)</span></span></code></pre></div>
<p><img src="Media_EP5/RestartR.png" width="400"></p>
</div>
<div class="section level3" number="4.5">
<h3 id="read-genbank-error">
<span class="header-section-number">4.5</span> read.GenBank
error<a class="anchor" aria-label="anchor" href="#read-genbank-error"></a>
</h3>
<p>So this is an anusual one that I don’t 100% understand. But, my guess
is that potentially GenBank is bloacking download requests from R over
maybe 376 accession numbers (sequences). Let me help you cheat the
system by breaking up the request. In this case we will:</p>
<ol style="list-style-type: decimal">
<li><p>Do our <em>entrez</em> search with too many sequences</p></li>
<li><p>Turn the search results into a tibble, break them into groups,
then into a list by group</p></li>
<li><p><em>[dplyr::]</em>pull out the column with the accession IDs
(<strong>NOTE</strong>: I’m just highlighting how dplyr lets you read
the code like a sentence!!)</p></li>
<li><p>Feed each list of IDs into <em>ape</em>’s function,
read.GenBank</p></li>
<li><p>Combine each returned read.GenBank result into one DNAbin
object!</p></li>
</ol>
<p>Now, the below is a reasonably extensive chunk of code. But, see how
with some experience in data management we have overcome a major issue
and done something that’s REALLY difficult in only a few lines of code
and now we can do it over and over again from now on! For this project
and for subsequent ones. Coding is powerful.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myCheekySearch</span> <span class="op">&lt;-</span> <span class="fu">rentrez</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/rentrez/reference/entrez_search.html" class="external-link">entrez_search</a></span><span class="op">(</span>db <span class="op">=</span> <span class="st">"nucleotide"</span>,</span>
<span>              term <span class="op">=</span> <span class="st">"(cytb[Gene Name]) AND (Apis[Organism])"</span>,</span>
<span>                <span class="co"># With this new search term, I can see 171 sequences, which is enough for me to</span></span>
<span>                <span class="co"># download, I think! So, I'll set this in retmax</span></span>
<span>              retmax <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspect the object and answer the question below</span></span>
<span><span class="va">myCheekySearch</span></span>
<span></span>
<span><span class="co"># What we actually need are the accession numbers form the myCheekySearch object. Let's turn this</span></span>
<span>  <span class="co"># vector into a list of vectors, each 200 long. I'm going to write it in the tidyverse because</span></span>
<span>  <span class="co"># it's easy for me, but it might not be the most concise way to do this.</span></span>
<span><span class="va">myCheekySearch_list</span> <span class="op">&lt;-</span> <span class="va">myCheekySearch</span><span class="op">$</span><span class="va">ids</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>accessionIDs <span class="op">=</span> <span class="va">.</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># Group by the row number and step size (200 IDs at once)</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span>group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/row_number.html" class="external-link">row_number</a></span><span class="op">(</span><span class="op">)</span><span class="op">/</span><span class="fl">200</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># Split the dataset up into a list by group</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_split.html" class="external-link">group_split</a></span><span class="op">(</span>.keep <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># You can check out any element of the list... let's look at the 2nd</span></span>
<span><span class="va">myCheekySearch_list</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span>  <span class="co"># Now we can apply the function across a list using lapply</span></span>
<span><span class="va">GenBank_list</span> <span class="op">&lt;-</span> <span class="va">myCheekySearch_list</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># FIrst, let's pass our list into lapply to extract (pull out) on the accessionIDs</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span></span>
<span>      <span class="co"># Remember, when you use the pipe, %&gt;%, you can call the data being passed in with "."</span></span>
<span>    X <span class="op">=</span> <span class="va">.</span>,</span>
<span>    FUN <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="va"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span>,</span>
<span>    <span class="va">accessionIDs</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span></span>
<span>    <span class="co"># The data to feed into the function</span></span>
<span>  X <span class="op">=</span> <span class="va">.</span>,</span>
<span>  FUN <span class="op">=</span> <span class="va">read.GenBank</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># This is called a "for loop" It is an EXTREMELY useful tool where you cant count from 1 to</span></span>
<span> <span class="co"># i iterations (in our case 1 to (i-1)). Don't worry too much about understanding it, but they</span></span>
<span>  <span class="co"># are extremely useful tools and can be quit equick to build for simple problems like below</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">GenBank_list</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="va">i</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">GenBank_loop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">GenBank_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="va">GenBank_list</span><span class="op">[[</span><span class="va">i</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>      <span class="co"># Make the attribute list for the first pair of DNAbin objects</span></span>
<span>    <span class="va">attributesList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>      names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">GenBank_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">names</span>, <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">GenBank_list</span><span class="op">[[</span><span class="va">i</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">names</span><span class="op">)</span>,</span>
<span>      class <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">GenBank_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">class</span>,</span>
<span>      description <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">GenBank_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">description</span>,</span>
<span>                      <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">GenBank_list</span><span class="op">[[</span><span class="va">i</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">description</span><span class="op">)</span>, </span>
<span>      species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">GenBank_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">species</span>,</span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">GenBank_list</span><span class="op">[[</span><span class="va">i</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">species</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>  <span class="op">}</span><span class="kw">else</span><span class="op">{</span></span>
<span>    <span class="va">GenBank_loop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">GenBank_loop</span>, <span class="va">GenBank_list</span><span class="op">[[</span><span class="va">i</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>            <span class="co"># Make the attribute list for the first pair of DNAbin objects</span></span>
<span>    <span class="va">attributesList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>      names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">attributesList</span><span class="op">$</span><span class="va">names</span>, <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">GenBank_list</span><span class="op">[[</span><span class="va">i</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">names</span><span class="op">)</span>,</span>
<span>      class <span class="op">=</span> <span class="va">attributesList</span><span class="op">$</span><span class="va">class</span>,</span>
<span>      description <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">attributesList</span><span class="op">$</span><span class="va">description</span>,</span>
<span>                      <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">GenBank_list</span><span class="op">[[</span><span class="va">i</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">description</span><span class="op">)</span>, </span>
<span>      species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">attributesList</span><span class="op">$</span><span class="va">species</span>,</span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">GenBank_list</span><span class="op">[[</span><span class="va">i</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">species</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span>  <span class="co"># We can then call this object whatever we want in R </span></span>
<span><span class="va">combinedGenBank</span> <span class="op">&lt;-</span> <span class="va">GenBank_loop</span></span>
<span></span>
<span>  <span class="co"># let's now add the attributes to this new object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">combinedGenBank</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">attributesList</span></span>
<span></span>
<span><span class="co"># This can be downloaded from GitHub as a function using the below code, if the code in Prac 3 fails due to permissions.</span></span>
<span><span class="co"># First of all, you will need to read in a function that is hosted on my GitHub</span></span>
<span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/jbdorey/EvolutionPracs/main/R/DoreyGenbank.R"</span>,</span>
<span>                     destfile <span class="op">=</span> <span class="st">"DoreyGenbank.R"</span>,</span>
<span>                     method<span class="op">=</span><span class="st">"curl"</span><span class="op">)</span></span>
<span>  <span class="co"># and then read in the function</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"DoreyGenbank.R"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2" number="5">
<h2 id="common-alignmenttree-issues">
<span class="header-section-number">5</span> Common alignment/tree
issues<a class="anchor" aria-label="anchor" href="#common-alignmenttree-issues"></a>
</h2>
<p><strong>Look, these issues are going to be hard to address here and
without being there to help you along, but let me try…</strong></p>
<div class="section level3" number="5.1">
<h3 id="incorrect-number-of-tips">
<span class="header-section-number">5.1</span> Incorrect number of
tips<a class="anchor" aria-label="anchor" href="#incorrect-number-of-tips"></a>
</h3>
<p>A common issue, especially when applying the code to your own taxa is
that the trees or alignments might say that you don’t have enough tips,
or maybe they are NULL, or maybe some other issue. You will need to
check your alginment to see that all of the names are there that you
expect! Check:</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB</span><span class="op">)</span></span></code></pre>
<p>Or</p>
<pre><code><span><span class="va">theChosenSeqs</span></span></code></pre>
<p>If there are too few names, or missing names then something has gone
wrong! Keep running the R objects that feed into each other until you
find where things change FROM WHAT YOU EXPECTED!! E.g.,</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB</span><span class="op">)</span></span>
<span><span class="va">theChosenSeqs</span></span>
<span><span class="va">genBankSummaryTibble</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">DarwinFinches_cytB</span><span class="op">)</span></span></code></pre>
<p>Quite often, the problem actually came from reading in the fasta file
itself! See the next point</p>
</div>
<div class="section level3" number="5.2">
<h3 id="cant-see-the-msar-alignment-box">
<span class="header-section-number">5.2</span> Can’t see the msaR
alignment box<a class="anchor" aria-label="anchor" href="#cant-see-the-msar-alignment-box"></a>
</h3>
<p>This is a weird issue on some PC laptops. I’m sorry to say that I
don’t know why and I don’t use PC to find out. But, I did find a
work-around that worked for some people. As with Figure 2 below, you can
select “Export” in the “Viewer” tab on the bottom right hand side of
RStudio and save as a Web Page. This can then be opened in your internet
browser (hopefully).</p>
<p><strong>REMEMBER: this is an interactive plot so scroll across and
see the alignment!</strong></p>
<p><img src="Media_EP5/msaR.png"></p>
</div>
<div class="section level3" number="5.3">
<h3 id="reading-in-a-fasta-from-a--fasta-file-vs-from-read-genbank">
<span class="header-section-number">5.3</span> Reading in a fasta
from a .fasta file vs from read.GenBank<a class="anchor" aria-label="anchor" href="#reading-in-a-fasta-from-a--fasta-file-vs-from-read-genbank"></a>
</h3>
<p>This is an important one and one where many people have had troubles.
Let me try to make this clear.</p>
<ol style="list-style-type: decimal">
<li>If you read in your data using <code><a href="https://rdrr.io/pkg/ape/man/read.GenBank.html" class="external-link">ape::read.GenBank()</a></code> then
R will automatically download a bunch of really useful attributes with
the data! The is why we can extract these data from
<code>attributes(youGenBank_data)</code> and turn it into the
<code>genBankSummaryTibble</code> Go and have a look for that in the
<strong>Prac 3 manual</strong>.</li>
<li>If you read your data in from a <strong>.fasta</strong> file that’s
totally fine BUT remember that fasta files contain only TWO bits of
information. They contain:
<ul>
<li>The name of your sequence</li>
<li>The sequence itself</li>
</ul>
</li>
</ol>
<p>Quite often the name of the sequence will have the accession ID, the
species name, and then a bunch of info including the gene name and such.
One easy way to get around this problem is to <strong>make a
copy</strong> of this file where you only include the species names
themselves, and remove the rest of that name part. We actually only did
all of that data management to have a nicer tree with shorter names!</p>
<p>Also, see point <strong>read.GenBank error</strong> above for an
alternative way to read from GenBank; if you had that problem! I may
have told you to just go and grab your sequences from GenBank directly
and using the accession numbers with `ape::read.GenBank()``.</p>
</div>
<div class="section level3" number="5.4">
<h3 id="problem-with-dnabin-object-or-type-of-dna-object">
<span class="header-section-number">5.4</span> Problem with DNAbin
object or type of DNA object<a class="anchor" aria-label="anchor" href="#problem-with-dnabin-object-or-type-of-dna-object"></a>
</h3>
<p>Some people have had issues feeding in DNAbin objects or other DNA
object types into functions. Remember, that I asked you to look at the
<code>class(ofYourObjects)</code>. Different functions are written to
handle certain types of data and I have written this code the same — I
expect a certain type of data to come into my code. Other types might be
fine, but they might require a chance in the code to deal with this!</p>
<p>Some common issues have included:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Using ChatGPT to help you with your code</strong>. For your
code (NOT YOUR WRITING), I will not discourage you from using ChatGPT,
but let’s be clear about one thing. ChatGPT can be a bit of an idiot. It
does not know the context of the rest of the code or what you are doing
and it will provide code that reflects this. For example, it might ask
you to read in your fasta using the below from the Biostrings package.
This might be fine if the rest of the code was designed with
<code>XStringSet</code> objects in mind. It was not</li>
</ol>
<pre><code><span><span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="op">)</span></span></code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Similar to the above you may realise that sometimes we read in fasta
objects in different ways! For example, we can read in the <strong>exact
same fasta file</strong> with two different packages and get different
objects, the first of the below being a DNAbin object and the second
coming out as a text string that we need to manipulate for use. We do
different things in different parts of the prac, so take your time and
have a look at what we usually feed into each section.</li>
</ol>
<p>Reads in as a DNAbin object:</p>
<pre><code><span><span class="va">alignedDNAbin</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.dna.html" class="external-link">read.FASTA</a></span><span class="op">(</span><span class="st">"alignedDNAbin.fasta"</span><span class="op">)</span> </span></code></pre>
<p>Reads in as a text string:</p>
<pre><code><span><span class="va">finchSpecies_fasta</span> <span class="op">&lt;-</span> <span class="fu">msaR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/msaR/man/as.fasta.html" class="external-link">as.fasta</a></span><span class="op">(</span><span class="va">finchSpecies</span><span class="op">)</span></span></code></pre>
<ol start="3" style="list-style-type: decimal">
<li>We also saved and read our fasta file as an <strong>.rds</strong>
data using <code><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS()</a></code> or <code><a href="https://rdrr.io/pkg/BiocGenerics/man/saveRDS.html" class="external-link">saveRDS()</a></code>. Why did we
do this instead of saving to a fasta file? Becuase, when we save a
DNAbin object this way we can keep the associated
<code>attributes</code> with it! Remember, fasta files only keep the
<em>&gt;name</em> and the <em>sequence</em>.</li>
</ol>
</div>
<div class="section level3" number="5.5">
<h3 id="random-tree-error">
<span class="header-section-number">5.5</span> Random tree
error<a class="anchor" aria-label="anchor" href="#random-tree-error"></a>
</h3>
<p>I will also point out that some people had trouble building their
trees because <strong>they did not run all of the preceding
code</strong>. So… go ahead and make sure you’ve run everything</p>
</div>
<div class="section level3" number="5.6">
<h3 id="msamsa-going-really-slow">
<span class="header-section-number">5.6</span> msa::msa going really
slow<a class="anchor" aria-label="anchor" href="#msamsa-going-really-slow"></a>
</h3>
<p>Look guys, I’ll start by saying that R can do some amazing things
with complex datasets and sometimes we just need to be patient while it
thinks. That said, many people are reading in whole mitogenomes
(16-18,000 basepairs of data per sequence). Remember, we talked about
this in Prac 2. I’m not going to stop you from aligning and building a
tree from mitogenomes but, depending on your computer, it will take a
lot longer. So have a think about if you want to just get one gene to
build your tree.</p>
</div>
<div class="section level3" number="5.7">
<h3 id="cannot-find-the-function-getymult">
<span class="header-section-number">5.7</span> Cannot find the
function “getYmult()”<a class="anchor" aria-label="anchor" href="#cannot-find-the-function-getymult"></a>
</h3>
<p>I was hoping that in this case <em>phytools</em> would give you a
helpful error about which package you needed to install for this
function. It does not. Regardless, a simple google of “getYmult()” would
return that it comes from the R package <em>plotrix</em> which is a
suggested (NOT COMPULSARY) dependency for <em>phytools</em>. Simply
install <em>plotrix</em>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"plotrix"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2" number="6">
<h2 id="data-issues">
<span class="header-section-number">6</span> Data issues<a class="anchor" aria-label="anchor" href="#data-issues"></a>
</h2>
<p>I have not had too many data issues yet. Remember, that I have
actually taught you and given you the code for a lot of data management!
But, you may also use Excel or similar if you’d prefer — look at the
marking guidelines and you’ll see that I’m not marking you on which
programs you use <em>per se</em>. Just be certain that you can meet
those criteria to get full marks (I’m happy to give out full marks)!</p>
<div class="section level3" number="6.1">
<h3 id="objects-are-different-length">
<span class="header-section-number">6.1</span> Objects are different
length<a class="anchor" aria-label="anchor" href="#objects-are-different-length"></a>
</h3>
<p>This is a common issues where you can’t combine two objects in R
because one is longer/shorter than the other. Probably, you have mixed
up code, missed a step or similar. In this case you need to track back
the problem by running each object that has led up to that point until
you find the number that does not make sense! Maybe, for example, you
read in a fasta file without attributes and then you only have your
outgroup name and not the ingroup species names! Let me give you an
illustrative example below.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Let's say that we make two simple data frames </span></span>
<span><span class="va">FirstDataFrame</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>column1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">SecondDataFrame</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>column1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"b"</span>, <span class="st">"c"</span>, <span class="st">"d"</span>, <span class="st">"e"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Can you already see the problem with joining these datasets? Let's have a look...</span></span>
<span>  <span class="co"># how long is each dataset?</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">FirstDataFrame</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 6</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">SecondDataFrame</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 5</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># They are different lengths! So, what happens when we do a very simple merge?</span></span>
<span><span class="va">failedMerge</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html" class="external-link">bind_cols</a></span><span class="op">(</span><span class="va">FirstDataFrame</span>, <span class="va">SecondDataFrame</span><span class="op">)</span></span></code></pre></div>
<p>You’ll get something like the below error. You will need to find the
problem and change the sizes so that they match! Alternatively, you can
look at how we used <code><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">dplyr::left_join()</a></code> in Prac 4; but this
depends on what you’re trying to do and on the error.</p>
<pre><code>Error in `dplyr::bind_cols()`:
! Can't recycle `..1` (size 6) to match `..2` (size 5).
Backtrace:
 1. dplyr::bind_cols(FirstDataFrame, SecondDataFrame)</code></pre>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James B. Dorey.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
