<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-gb">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Evolution prac 4 — Macroevolutionary questions! • Evolution Practicals</title>
<!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Evolution prac 4 — Macroevolutionary questions!">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Evolution Practicals</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html" aria-label="Go to package home"><span class="fa fa-home"></span></a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Available pages"><span class="fa fas fa-book"></span> Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/PhyloPrac1.html">Evolution Prac 1 — Getting ready!</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac2.html">Evolution Prac 2 — Access genetic data from GenBank</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac3.html">Evolution Prac 3 — Aligning genetic data and building phylogenies in R</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac4.html">Evolution prac 4 — Macroevolutionary questions!</a></li>
    <li><a class="dropdown-item" href="../articles/PhyloPrac5.html">Evolution prac 5 — Tips and pitfalls</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">


<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Evolution prac 4 — Macroevolutionary questions!</h1>
                        <h4 data-toc-skip class="author">Dr James B
Dorey and Dr Damien Esquerré, University of Wollongong</h4>
            
            <h4 data-toc-skip class="date">Version 2025-07-17; practical
2025-08-22</h4>
      

      <div class="d-none name"><code>PhyloPrac4.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      Today we will bee starting to use phylogenetic trees to ask and
      try to answer interesting questions. These questions will be up to
      you, for your assignments, but could relate to ecological,
      phenotypic, or biogeographical information. Hopefully, you’ve got
      a pretty good idea of a taxa of interest and a question that you
      want to ask about it. You might have the question and see if you
      can find the data, or you might find some data and see if you can
      interrogate it in an interesting way. Get creative and get
      thinking; these are timeless skills!<br><strong>Be certain to save your R script from this week to help
      you with your assignment!</strong><br><strong>I have a strong suspicion that you will come across more
      package issues this week. But, I am also quite confident that, by
      reading the error and applying your R knowledge you will be able
      to solve those issues! </strong>
    </div>
    
<div class="alert alert-info">
<p><strong> Attention:</strong> <br><strong>Once you have completed all
of the questions</strong> and <strong>shown your demonstrator or
Lecturer</strong> the answers and the outputs, you may leave. Or, feel
free to hang around, play, or help your classmates.</p>
</div>
<div class="section level2" number="1">
<h2 id="script-preparation">
<span class="header-section-number">1</span> Script preparation<a class="anchor" aria-label="anchor" href="#script-preparation"></a>
</h2>
<div class="section level3" number="1.1">
<h3 id="working-directory">
<span class="header-section-number">1.1</span> Working
directory<a class="anchor" aria-label="anchor" href="#working-directory"></a>
</h3>
<p>Just like last prac, we will set our working directory. This will,
likely, be the same directory as you used last week as it’s all one
project.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Set the RootPath to Prac1 folder</span></span>
<span><span class="va">RootPath</span> <span class="op">&lt;-</span> <span class="st">"YourFolderPathHere"</span></span>
<span>  <span class="co"># You can then set this as the project's working directory. </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="va">RootPath</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3" number="1.2">
<h3 id="install-packages">
<span class="header-section-number">1.2</span> Install packages<a class="anchor" aria-label="anchor" href="#install-packages"></a>
</h3>
<p>We should already have the packages from last week installed.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#  A tidyverse package for reading in Excel sheets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"readxl"</span>,</span>
<span>              repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Media_EP4/readxl.png" width="100"></p>
<p><br></p>
</div>
<div class="section level3" number="1.3">
<h3 id="load-packages">
<span class="header-section-number">1.3</span> Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<p>Once again, we will need to load all of our packages into R.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dplyr"</span>, <span class="st">"magrittr"</span>, <span class="st">"ape"</span>, <span class="st">"rentrez"</span>, <span class="st">"tidyr"</span>, <span class="st">"stringr"</span><span class="op">)</span>, </span>
<span>       <span class="va">library</span>, character.only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2" number="2">
<h2 id="load-and-prepare-genetic-and-trait-data">
<span class="header-section-number">2</span> Load and prepare
genetic and trait data<a class="anchor" aria-label="anchor" href="#load-and-prepare-genetic-and-trait-data"></a>
</h2>
<div class="section level3" number="2.1">
<h3 id="genetic-alignment-from-last-week">
<span class="header-section-number">2.1</span> Genetic alignment
(from last week)<a class="anchor" aria-label="anchor" href="#genetic-alignment-from-last-week"></a>
</h3>
<p>Let us start off about where we were yesterday by loading in the
genetic alignment that we saved last week. You can download last week’s
alignment using the code below as I have hosted the data online.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/jbdorey/EvolutionPracs/main/EPrac4/alignedDNAbin.fasta"</span>,</span>
<span>                     destfile <span class="op">=</span> <span class="st">"alignedDNAbin.fasta"</span>,</span>
<span>                     method<span class="op">=</span><span class="st">"curl"</span><span class="op">)</span></span></code></pre></div>
<p>Okay, we have downloaded these data and you can go ahead and look in
your working directory (<em>if you’re not sure type “getwd()” into R to
see where this is</em>). Let’s go ahead and read it into R.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Use ape to read in the aligned sequences</span></span>
<span><span class="va">alignedDNAbin</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.dna.html" class="external-link">read.FASTA</a></span><span class="op">(</span><span class="st">"alignedDNAbin.fasta"</span><span class="op">)</span> </span>
<span>  <span class="co"># Let's also simplify our sequence names (this'll make things easier and </span></span>
<span>  <span class="co"># look better in our tree!)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">alignedDNAbin</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">alignedDNAbin</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove_all</a></span><span class="op">(</span><span class="st">"HQ[0-9]|[0-9]"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace_all</a></span><span class="op">(</span><span class="st">"_"</span>, <span class="st">" "</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_trim.html" class="external-link">str_squish</a></span><span class="op">(</span><span class="op">)</span> </span>
<span></span>
<span>  <span class="co"># Let's also make the Maximum Likelihood (ML) tree from last week  </span></span>
<span><span class="va">finchesML</span> <span class="op">&lt;-</span> <span class="va">alignedDNAbin</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/as.phyDat.html" class="external-link">phyDat</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/pml_bb.html" class="external-link">pml_bb</a></span><span class="op">(</span>model <span class="op">=</span> <span class="st">"HKY"</span>, </span>
<span>                   control <span class="op">=</span> <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/pml.control.html" class="external-link">pml.control</a></span><span class="op">(</span>trace <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/bootstrap.pml.html" class="external-link">bootstrap.pml</a></span><span class="op">(</span>bs <span class="op">=</span> <span class="fl">100</span>, optNni <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                          method <span class="op">=</span> <span class="st">"ultrametric"</span>, </span>
<span>                          control <span class="op">=</span> <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/pml.control.html" class="external-link">pml.control</a></span><span class="op">(</span>trace <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># plot the tree and write it to the object "treeML"</span></span>
<span><span class="op">(</span><span class="va">treeML</span> <span class="op">&lt;-</span> <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/plotBS.html" class="external-link">plotBS</a></span><span class="op">(</span></span>
<span>  tree <span class="op">=</span> <span class="va">finchesML</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>      <span class="co"># Find the Maximum clade credibility tree</span></span>
<span>    <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/maxCladeCred.html" class="external-link">maxCladeCred</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>      <span class="co"># We can root the tree at this point</span></span>
<span>    <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/root.html" class="external-link">root</a></span><span class="op">(</span>outgroup <span class="op">=</span> <span class="st">"Asemospiza obscura"</span><span class="op">)</span>,</span>
<span>  p <span class="op">=</span> <span class="fl">10</span>, type<span class="op">=</span><span class="st">"p"</span>, digits<span class="op">=</span><span class="fl">2</span>, main<span class="op">=</span><span class="st">"Ultrafast bootstrap"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="PhyloPrac4_files/figure-html/readFasta-1.png" width="672"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Phylogenetic tree with 15 tips and 11 internal nodes.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Tip labels:</span></span>
<span><span class="co">#&gt;   Camarhynchus parvulus, Camarhynchus pauper, Camarhynchus psittacula, Camarhynchus heliobates, Geospiza scandens, Geospiza conirostris, ...</span></span>
<span><span class="co">#&gt; Node labels:</span></span>
<span><span class="co">#&gt;   0.81, NA, 0.26, 0.31, 0.3, 0.41, ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Unrooted; includes branch length(s).</span></span></code></pre>
</div>
<div class="section level3" number="2.2">
<h3 id="finch-trait-data">
<span class="header-section-number">2.2</span> Finch trait data<a class="anchor" aria-label="anchor" href="#finch-trait-data"></a>
</h3>
<p>The reason that we’ve been doing finches this whole time is that I
know it’s easy enough to get some finch data from the internet! We will
download data from a Dryad database that digitised 6,500 specimens that
were measured by David Lack! A treasure trove of data, so thanks to
David for making them available!</p>
<p>You can download the data from this link — <a href="https://datadryad.org/stash/dataset/doi:10.5061/dryad.150" class="external-link uri">https://datadryad.org/stash/dataset/doi:10.5061/dryad.150</a>
— under <strong>Download full dataset</strong> on the top right hand
bar. Save these to your working directory. We will read in
“morphLack.xls”</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/jbdorey/EvolutionPracs/main/EPrac4/morphLack.xls"</span>,</span>
<span>                     destfile <span class="op">=</span> <span class="st">"morphLack.xls"</span>,</span>
<span>                     method<span class="op">=</span><span class="st">"curl"</span><span class="op">)</span></span>
<span><span class="va">finchData</span> <span class="op">&lt;-</span> <span class="fu">readxl</span><span class="fu">::</span><span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_excel</a></span><span class="op">(</span><span class="st">"morphLack.xls"</span>, sheet <span class="op">=</span> <span class="st">"morphLack.TAB"</span>, guess_max <span class="op">=</span> <span class="fl">7000</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We can read these data into R very easily using readxl</span></span>
<span><span class="va">finchData</span> <span class="op">&lt;-</span> <span class="fu">readxl</span><span class="fu">::</span><span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_excel</a></span><span class="op">(</span><span class="st">"morphLack.xls"</span>,</span>
<span>                         sheet <span class="op">=</span> <span class="st">"morphLack.TAB"</span>,</span>
<span>                          <span class="co"># readxl needs to figure out what's in each column (text, </span></span>
<span>                          <span class="co"># numbers...), so let it read all rows to figure this out.</span></span>
<span>                         guess_max <span class="op">=</span> <span class="fl">7000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Have a look at this big data table:</span></span>
<span><span class="va">finchData</span></span></code></pre></div>
<p>This finch data table has 6,758 data observations across 20
variables. We can guess what some of the columns are:
<strong>IslandID</strong> = island (could be good for biogeography
analyses), <strong>BodyL</strong> = body length; <strong>WingL</strong>
= wing length; <strong>TailL</strong> = tail length;
<strong>BeakW</strong> = beak width; <strong>BeakH</strong> = beak
height; <strong>LBeakL</strong> = lower beak length;
<strong>UBeakL</strong> = upper beak length; <strong>TarsusL</strong> =
tarsus length; and I don’t know what <strong>N-UBkL</strong> is… oh
well. The point is that there are traits that we can start asking
macroevolutionary questions about. Maybe we can get at a classic and
look at beak size (don’t worry, I don’t think we’re treading on Darwin’s
toes).</p>
<p>One potential issue with this dataset, as impressive as it is, there
is a lot of missing data. But, that’s okay. Let us just use one of the
more-complete variables like beak height. The next thing that we might
consider is that there are loads of measurements and we only want one
per species. We could take the mean or median values per species.
Firstly, we can make a quick histogram of this variable to see if it is
normally distributed…</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Have a look at the histogram</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">finchData</span><span class="op">$</span><span class="va">BeakH</span><span class="op">)</span></span>
<span><span class="co"># I'd say that's right-skewed and so let's use a median value</span></span>
<span><span class="co"># Let's calculate the median for each species and put it into a new column</span></span>
<span><span class="va">finchSummary</span> <span class="op">&lt;-</span> <span class="va">finchData</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># There are two sets of names in the dataset, original and</span></span>
<span>    <span class="co"># updated, let's use the latter and put it in a new column</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>scientificName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">GenusL69</span>, <span class="va">SpeciesL69</span>, sep <span class="op">=</span> <span class="st">" "</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Group by scientificName to calculate the mean within groups</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">scientificName</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># let's get the median beak height per species</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>medianBH <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">BeakH</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                    <span class="co"># The range is the max minus the min values per species</span></span>
<span>                   rangeBH <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">BeakH</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">BeakH</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Have a look at these data by getting a quick summary</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">finchSummary</span><span class="op">)</span></span></code></pre></div>
<div class="alert alert-warning" role="alert">
<blockquote>
<p><strong>Q1:</strong> What is the range of the median beak height
values?</p>
</blockquote>
</div>
<p>I’ll show you, but don’t expect you to emulate this, that we can use
a <em>Shapiro-Wilk Normality Test</em> to statistically test if our data
significantly differs from a normal distribution</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># If p-value &gt; 0.05, we'd say that we can't reject the null hypothesis (normal</span></span>
<span>  <span class="co"># distribution)... This is close enough!</span></span>
<span><span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/shapiro.test.html" class="external-link">shapiro.test</a></span><span class="op">(</span><span class="va">finchSummary</span><span class="op">$</span><span class="va">medianBH</span><span class="op">)</span></span>
<span>  <span class="co"># The range is NOT normally distributed</span></span>
<span><span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/shapiro.test.html" class="external-link">shapiro.test</a></span><span class="op">(</span><span class="va">finchSummary</span><span class="op">$</span><span class="va">rangeBH</span><span class="op">)</span></span>
<span>  <span class="co"># But, if we log-transform this variable, it is normal!</span></span>
<span><span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/shapiro.test.html" class="external-link">shapiro.test</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">finchSummary</span><span class="op">$</span><span class="va">rangeBH</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Media_EP4/normalMeme.jpeg" height="250"></p>
</div>
<div class="section level3" number="2.3">
<h3 id="match-genetic-and-trait-data">
<span class="header-section-number">2.3</span> Match genetic and
trait data<a class="anchor" aria-label="anchor" href="#match-genetic-and-trait-data"></a>
</h3>
<p>We have two different datasets that we want to work together with, so
let’s see how well they overlap with one another and then start making
them work together.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Let us have a look at how well our taxon names match between the genes</span></span>
<span><span class="co"># and the trait data</span></span>
<span>  <span class="co"># Extract the DNA names</span></span>
<span><span class="va">dnaNames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">alignedDNAbin</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Turn them into a data frame (tibble)</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>scientificName <span class="op">=</span> <span class="va">.</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># We can perform a "left_join" to put these tibbles together and see what's</span></span>
<span>  <span class="co"># missing! </span></span>
<span><span class="va">compareNameTable</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">dnaNames</span>, <span class="va">finchSummary</span>, by <span class="op">=</span> <span class="st">"scientificName"</span><span class="op">)</span></span></code></pre></div>
<p>Looks like we’ve got pretty good overlap! The trait data is only
missing three of our genetic species and one of those is the outgroup.
In your own project, you could absolutely go hunting for that extra data
but, for our purposes, I’m pretty happy with that result. We can look at
trait evolution across 12 species with these data.</p>
<p>Now, let’s <em>prune</em> our tree to make sure that the data
overlap.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># We can drop the tips on our tree that don't have trait data associated with them.</span></span>
<span><span class="va">treeML_dropped</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/drop.tip.html" class="external-link">drop.tip</a></span><span class="op">(</span><span class="va">treeML</span>,</span>
<span>                      tip <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Geospiza septentrionalis"</span>, <span class="st">"Certhidea fusca"</span>, </span>
<span>                              <span class="st">"Asemospiza obscura"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Let's only select the traits for those species in the tree</span></span>
<span><span class="va">finchSummary</span> <span class="op">&lt;-</span> <span class="va">finchSummary</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># We can find the names %in% the tree tip labels slot and filter for those</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">scientificName</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">treeML_dropped</span><span class="op">$</span><span class="va">tip.label</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># let's also add the log transformation to normalise our range data</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>log_rangeBH <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">rangeBH</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now we have matching datasets. With our matching datasets, we can
start to answer some hypotheses! For the sections that are coming, I’ll
be using the book, <strong>Phylogenetic Comparitive Methods in R by Liam
Revell &amp; Luke Harmon</strong>, as a reference. It’s a great book and
Liam is also the author of a fantastic R package,
<strong>phytools</strong>.</p>
</div>
</div>
<div class="section level2" number="3">
<h2 id="phylogenetically-independent-contrasts">
<span class="header-section-number">3</span> Phylogenetically
independent contrasts<a class="anchor" aria-label="anchor" href="#phylogenetically-independent-contrasts"></a>
</h2>
<p>In comparative biology, we will often be asking questions about
traits and if they are related to one another. For example, do bigger
animals have larger home ranges? This is a relatively easy question to
ask and answer as we can simply use regression to see the correlation
between those two traits. However, a key assumption of linear regression
(like many other analyses) is that data points are independent. The
problem is that more-closely related taxa are more likely to share
similar traits and so <strong>cannot be considered independent</strong>
from one another.</p>
<p><a href="https://www.jstor.org/stable/2461605" class="external-link">Joseph Felsenstein
(1985)</a> pointed out this issue that trait correlations could,
especially in worst case scenarios (Fig. 1), could lead to misleading
results. He proposed then that, if traits evolve randomly (by Brownian
motion) in each lineage after each node, you could correct for this
bias. This correction technique is termed <strong>phylogenetically
independent contrasts</strong>. The mathematics behind this correction,
I’ll admit, is a little beyond me. However, the use of these tools is
not.</p>
<div class="float">
<img src="Media_EP4/Felsenstein1985_plots.png" alt="Figures 5 and 7 from Felenstein, 1985 showing a hypothetical phylogeny and how the traits of those taxa might, through shared ancestry and nonindependence, bias a regular regression analysis."><div class="figcaption">Figures 5 and 7 from Felenstein, 1985 showing a
hypothetical phylogeny and how the traits of those taxa might, through
shared ancestry and nonindependence, bias a regular regression
analysis.</div>
</div>
<p>Before moving on, let’s make a simple hypothesis. <strong>Finches
with larger beak heights will have larger beak height ranges than
finches with smaller beak heights</strong>. I have no idea if this
hypothesis has been asked before, but you can see that I have provided a
TESTABLE hypothesis that will serve as a great bounding board for some
tests.</p>
<div class="section level3" number="3.1">
<h3 id="regular-regression">
<span class="header-section-number">3.1</span> Regular
regression<a class="anchor" aria-label="anchor" href="#regular-regression"></a>
</h3>
<p>Let’s begin by using regular-old linear regression to test our
hypothesis. Keep in mind that, as of writing this, I don’t know what the
answer will be. For this we can ignore the phylogeny</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># We can start by just plotting these two variables to see how they look</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">finchSummary</span><span class="op">$</span><span class="va">medianBH</span>, <span class="va">finchSummary</span><span class="op">$</span><span class="va">log_rangeBH</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># We should then run a linear regression on the same data</span></span>
<span>  <span class="co"># Firstly, let's make the model!</span></span>
<span><span class="va">linear_BH</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">finchSummary</span>,</span>
<span>                       formula <span class="op">=</span> <span class="va">log_rangeBH</span> <span class="op">~</span> <span class="va">medianBH</span><span class="op">)</span></span>
<span>  <span class="co"># Have a peak and you might see where the y-intercept and slope of a linear model</span></span>
<span>  <span class="co"># (y = mx + c). In the model, c = Intercept; m = log_rangeBH</span></span>
<span><span class="va">linear_BH</span></span>
<span></span>
<span>  <span class="co"># Now, let's test our hypothesis!</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">linear_BH</span><span class="op">)</span></span></code></pre></div>
<div class="alert alert-warning" role="alert">
<blockquote>
<p><strong>Q2:</strong> Ignoring phylogeny, are median beak height and
log range of beak height significantly correlated? What’s the
significance level (p-value) for the model?</p>
</blockquote>
</div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Like I mentioned, the linear model is jsut a line; so we can also plot that line!</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">finchSummary</span><span class="op">$</span><span class="va">medianBH</span>, <span class="va">finchSummary</span><span class="op">$</span><span class="va">log_rangeBH</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="va">linear_BH</span>, col <span class="op">=</span> <span class="st">"red"</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3" number="3.2">
<h3 id="phylogentically-independent-contrasts-pic">
<span class="header-section-number">3.2</span> Phylogentically
independent contrasts (PIC)<a class="anchor" aria-label="anchor" href="#phylogentically-independent-contrasts-pic"></a>
</h3>
<p>Okay, now we’ve checked using regular linear regression, let’s use
our phylogenetic correction. We will use <strong>ape</strong>’s
<code><a href="https://rdrr.io/pkg/ape/man/pic.html" class="external-link">pic()</a></code> function to do this.</p>
<p><strong>Note:</strong> We actually have a slight problem here, most
<em>Camarhynchus</em> species sequences are identical for cytB. The same
is maybe true for <em>Geospiza magnirostris</em> and <em>Geospiza
conirostris</em>. These are all visible as <em>zero branch lengths</em>
in the phylogeny. We must then fix this (or drop those taxa). Don’t
worry too much about the first bit of code, I’ll basically add a
<em>small</em> branch length for these species and then make them split,
using <code><a href="https://rdrr.io/pkg/ape/man/multi2di.html" class="external-link">ape::multi2di()</a></code>. It will add a little noise to the
data, but should be fine as a demonstration; we could also fix this by
adding more genetic data or choosing a different gene with more
variation.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Make a new tree to not save-over our "real" tree</span></span>
<span><span class="va">treeML_fixed</span> <span class="op">&lt;-</span> <span class="va">treeML_dropped</span></span>
<span>  <span class="co"># Add a small branch length (0.001) to those with an exactly zero length (0);</span></span>
<span>  <span class="co"># Looks like some are almost zero (0.00000001); we also want to fix these!</span></span>
<span><span class="va">treeML_fixed</span><span class="op">$</span><span class="va">edge.length</span> <span class="op">&lt;-</span> <span class="va">treeML_fixed</span><span class="op">$</span><span class="va">edge.length</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">.</span> <span class="op">&lt;</span> <span class="fl">0.001</span>, <span class="fl">0.001</span>, <span class="va">.</span><span class="op">)</span></span>
<span>  <span class="co"># Make these new branch lengths bifurcate (split)</span></span>
<span><span class="va">treeML_fixed</span> <span class="op">&lt;-</span> <span class="va">treeML_fixed</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/multi2di.html" class="external-link">multi2di</a></span><span class="op">(</span>random <span class="op">=</span> <span class="cn">TRUE</span>,tol <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Feel free to compare how the original and changed trees look!</span></span>
<span><span class="co"># Create a side-by-side layout</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># Then make each plot!</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">treeML_dropped</span>, main <span class="op">=</span> <span class="st">"Original plot"</span>, sub <span class="op">=</span> <span class="st">"node numbers shown for illustration"</span>, </span>
<span>     col.sub <span class="op">=</span> <span class="st">"darkred"</span>, cex <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span>  <span class="co"># We can also plot the node numbers that will relate to the pic values coming up!</span></span>
<span><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/nodelabels.html" class="external-link">nodelabels</a></span><span class="op">(</span><span class="va">treeML_dropped</span><span class="op">$</span><span class="va">node.label</span>, bg <span class="op">=</span> <span class="st">"white"</span>, frame <span class="op">=</span> <span class="st">"circle"</span>, cex <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">treeML_fixed</span>, main <span class="op">=</span> <span class="st">"New plot"</span>, sub <span class="op">=</span> <span class="st">"now with tiny branches added"</span>, </span>
<span>     col.sub <span class="op">=</span> <span class="st">"darkred"</span>, cex <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="PhyloPrac4_files/figure-html/picCorrection-1.png" width="672"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Return R's plots to a single one! </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Okay, let’s now do our PIC! We need to run our pic regression through
zero and so will add “+0” to our model. The reason for this is that the
direction of rotation for any node (right or left) is completely
arbitrary and so is the direction of subtraction (pic values can be - or
+; see <code>pic_logRangeBH</code>). So our pic regression should go
through zero and have no intercept term. <em>(Running without this term
gives a VERY similar answer anyway.)</em></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Let's calculate PIC for log range BH and for median BH</span></span>
<span><span class="va">pic_logRangeBH</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/pic.html" class="external-link">pic</a></span><span class="op">(</span><span class="va">finchSummary</span><span class="op">$</span><span class="va">log_rangeBH</span>, </span>
<span>                 <span class="va">treeML_fixed</span><span class="op">)</span></span>
<span><span class="va">pic_medianBH</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/pic.html" class="external-link">pic</a></span><span class="op">(</span><span class="va">finchSummary</span><span class="op">$</span><span class="va">medianBH</span>, </span>
<span>                 <span class="va">treeML_fixed</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># We can use these PIC's to run a new linear regression! </span></span>
<span><span class="va">picLM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="va">pic_medianBH</span> <span class="op">~</span> <span class="va">pic_logRangeBH</span> <span class="op">+</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Let's test this new hypothesis</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">picLM</span><span class="op">)</span></span></code></pre></div>
<div class="alert alert-warning" role="alert">
<blockquote>
<p><strong>Q3:</strong> Do you come to the same or a different
conclusion, in regards to our hypothesis, given this pic analysis?
What’s the p-value of this model?</p>
</blockquote>
</div>
<p>We can plot this model as well!</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Make the plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">pic_logRangeBH</span>, <span class="va">pic_medianBH</span><span class="op">)</span></span>
<span>  <span class="co"># Add in some dashed lines through zero on the horizontal and vertical</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fl">0</span>, lty <span class="op">=</span> <span class="st">"dotted"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="fl">0</span>, lty <span class="op">=</span> <span class="st">"dotted"</span><span class="op">)</span></span>
<span>  <span class="co"># Add in the model regression.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="va">picLM</span>, col <span class="op">=</span> <span class="st">"red"</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p>Listen, I’ve given you a lot to think about here. I’d be very pleased
if you’ve come this far in our prac. What follows is also a very useful
tool in macroevolutionary analyses. I’d encourage you to at least have a
look at the outputs and try to make it through the <em>continuous
variables</em> part. <em>Categorical variables</em> get a little more
complicated.</p>
</div>
</div>
<div class="section level2" number="4">
<h2 id="ancestral-state-reconstruction">
<span class="header-section-number">4</span> Ancestral state
reconstruction<a class="anchor" aria-label="anchor" href="#ancestral-state-reconstruction"></a>
</h2>
<p>We will now use an awesome package called <strong>phytools</strong>,
which was written by Liam Revell, who also wrote <strong>Phylogenetic
Comparitive Methods in R</strong>. We can install that now.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"phytools"</span>,</span>
<span>              repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/liamrevell/phytools" class="external-link">phytools</a></span><span class="op">)</span></span></code></pre></div>
<p><img src="Media_EP4/phytools.png" width="100"></p>
<p><br></p>
<p>We can reconstruct the probable state of a variable (continuous or
categorical) along the phylogeny using <em>ancestral state
reconstructions</em>. We will do just that!</p>
<div class="section level4" number="4.0.1">
<h4 id="continuous-variable">
<span class="header-section-number">4.0.1</span> Continuous
variable<a class="anchor" aria-label="anchor" href="#continuous-variable"></a>
</h4>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get a vector with the median beak heights and the scientific names</span></span>
<span><span class="va">medianVector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">finchSummary</span><span class="op">$</span><span class="va">medianBH</span>, <span class="va">finchSummary</span><span class="op">$</span><span class="va">scientificName</span><span class="op">)</span></span>
<span> </span>
<span><span class="co"># Use phytools to reconstruct ancestral states is the fastAnc function, which</span></span>
<span><span class="co"># estimates ancestral states using maximum likelihood</span></span>
<span><span class="va">ancMedian</span> <span class="op">&lt;-</span> <span class="fu">phytools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phytools/man/fastAnc.html" class="external-link">fastAnc</a></span><span class="op">(</span><span class="va">treeML_fixed</span>, <span class="va">medianVector</span>,</span>
<span>                  vars <span class="op">=</span> <span class="cn">TRUE</span>, CI <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># We can plot this tree using phytools and then add in the estimated values</span></span>
<span><span class="co"># At the nodes to see how the birds have evolved in this trait</span></span>
<span>  <span class="co"># Plot the tree using phytools</span></span>
<span><span class="fu">phytools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phytools/man/plotTree.html" class="external-link">plotTree</a></span><span class="op">(</span><span class="va">treeML_fixed</span>, ftype <span class="op">=</span> <span class="st">"i"</span>, fsize <span class="op">=</span> <span class="fl">0.5</span>, lwd <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="co"># Add in the nodes</span></span>
<span><span class="fu">phytools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phytools/man/labelnodes.html" class="external-link">labelnodes</a></span><span class="op">(</span>text <span class="op">=</span> <span class="va">ancMedian</span><span class="op">$</span><span class="va">ace</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                     node <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ancMedian</span><span class="op">$</span><span class="va">ace</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                     interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                     shape <span class="op">=</span> <span class="st">"circle"</span>,</span>
<span>                     cex <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="PhyloPrac4_files/figure-html/fastAnc-1.png" width="672"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  </span>
<span><span class="co"># We can also use phytools to map this evolution as a continuous trait along</span></span>
<span><span class="co"># our tree branchs!</span></span>
<span><span class="va">plotMap</span> <span class="op">&lt;-</span> <span class="fu">phytools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phytools/man/contMap.html" class="external-link">contMap</a></span><span class="op">(</span><span class="va">treeML_fixed</span>, <span class="va">medianVector</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">phytools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phytools/man/setMap.html" class="external-link">setMap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"orange"</span>, <span class="st">"black"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Plot the tree</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">plotMap</span>, leg.txt <span class="op">=</span> <span class="st">"Median beak height (mm)"</span>, fsize <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="PhyloPrac4_files/figure-html/fastAnc-2.png" width="672"></p>
<p>Hey, look at that! There are very few species with high beaks, and
they are mostly grouped in one genus, <em>Geospiza</em>! That’s
cool.</p>
<div class="alert alert-warning" role="alert">
<blockquote>
<p><strong>Q4:</strong> From the tree, which species has the biggest
median beak height?</p>
</blockquote>
</div>
</div>
<div class="section level4" number="4.0.2">
<h4 id="categorical-variable">
<span class="header-section-number">4.0.2</span> Categorical
variable<a class="anchor" aria-label="anchor" href="#categorical-variable"></a>
</h4>
<p>We don’t necessarily have a great categorical data in our dataset.
However, we could make a continuous variable categorical, and so let’s
use beak height again. We can make a histogram and see that there are
some natural breaks in beak height (see below). Let’s split it into
small (&lt;10 mm) and large beaks (&gt;10 mm).</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># See the histogram of species median beak height</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">finchSummary</span><span class="op">$</span><span class="va">medianBH</span>, breaks <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="PhyloPrac4_files/figure-html/categorical%20reconstruction-1.png" width="672"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Make this into a categorical variable using an if-else statement</span></span>
<span><span class="va">finchSummary</span> <span class="op">&lt;-</span> <span class="va">finchSummary</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>beakCat <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">medianBH</span> <span class="op">&gt;</span> <span class="fl">10</span>,</span>
<span>      <span class="co"># IF medianBH &gt; 10:</span></span>
<span>    <span class="st">"Large"</span>, </span>
<span>      <span class="co"># ELSE medianBH &lt; or = 10:</span></span>
<span>    <span class="st">"Small"</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get a vector with the beak categories and the scientific names</span></span>
<span><span class="va">beakVector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">finchSummary</span><span class="op">$</span><span class="va">beakCat</span>, <span class="va">finchSummary</span><span class="op">$</span><span class="va">scientificName</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Just like we have done for our phylogenetic tree evolution models, we need to</span></span>
<span><span class="co"># choose a model of evolution (transition) for our categorical variable</span></span>
<span><span class="co"># In this case, it's a bit more of a pain and we need to build the models and then</span></span>
<span><span class="co"># test them ourselves</span></span>
<span>  <span class="co"># Equal rates between each = ER</span></span>
<span><span class="va">ER_fit</span> <span class="op">&lt;-</span> <span class="fu">phytools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phytools/man/fitMk.html" class="external-link">fitMk</a></span><span class="op">(</span><span class="va">treeML_fixed</span>, </span>
<span>                        <span class="va">beakVector</span>,</span>
<span>                        model <span class="op">=</span> <span class="st">"ER"</span><span class="op">)</span></span>
<span>  <span class="co"># All rates different (ARD) means that all transition rates can be different</span></span>
<span><span class="va">ARD_fit</span> <span class="op">&lt;-</span> <span class="fu">phytools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phytools/man/fitMk.html" class="external-link">fitMk</a></span><span class="op">(</span><span class="va">treeML_fixed</span>, </span>
<span>                        <span class="va">beakVector</span>,</span>
<span>                        model <span class="op">=</span> <span class="st">"ARD"</span><span class="op">)</span></span>
<span>  <span class="co"># We could compare the fit of these models as below</span></span>
<span><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  model <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ER"</span>, <span class="st">"ARD"</span><span class="op">)</span>,</span>
<span>  LogL <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/logLik.html" class="external-link">logLik</a></span><span class="op">(</span><span class="va">ER_fit</span><span class="op">)</span>, <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/logLik.html" class="external-link">logLik</a></span><span class="op">(</span><span class="va">ARD_fit</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  aic <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/AIC.html" class="external-link">AIC</a></span><span class="op">(</span><span class="va">ER_fit</span><span class="op">)</span>, <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/AIC.html" class="external-link">AIC</a></span><span class="op">(</span><span class="va">ARD_fit</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 3</span></span></span>
<span><span class="co">#&gt;   model  LogL   aic</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ER    -<span style="color: #BB0000;">9.88</span>  21.8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ARD   -<span style="color: #BB0000;">9.88</span>  23.8</span></span>
<span></span>
<span><span class="co"># HOWEVER, there are only two states and so not much to compare! If there were more</span></span>
<span><span class="co"># states then you'd choose those with the lowest AIC and biggest Log likelihood.</span></span></code></pre></div>
<p>Now, let’s do something cool and calculate the probability that each
node is large or small using <em>marginal ancestral state
reconstruction</em>. We’ll need another package for this —
<strong>corHMM</strong>. <em>Sorry, it doesn’t have a logo :(</em></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install and load the package</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"corHMM"</span>,</span>
<span>              repos <span class="op">=</span> <span class="st">"http://cran.us.r-project.org"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">corHMM</span><span class="op">)</span></span></code></pre></div>
<p>You’re gonna hate me for this, but we need to turn our categories
into numbers (0, 1, 2, 3, … etcetera)! We can do this transformation in
text by converting it first into a factor and then into a number and
taking 1 away so that it starts at zero.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Modify the finch data to be numbered categorical with only species name</span></span>
<span><span class="co"># and that numbered factor in a DATA FRAME format</span></span>
<span><span class="va">funchNumbered</span> <span class="op">&lt;-</span> <span class="va">finchSummary</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">scientificName</span>, <span class="va">beakCat</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>beakCatNumbered <span class="op">=</span> <span class="va">beakCat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>  <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">!</span><span class="va">beakCat</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Undertake the marginal reconstruction. This is a pretty complex bit of code.</span></span>
<span><span class="co"># Please do examine ?corHMM for help.</span></span>
<span><span class="va">cor_margRecon</span> <span class="op">&lt;-</span> <span class="fu">corHMM</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/corHMM/man/corHMM.html" class="external-link">corHMM</a></span><span class="op">(</span></span>
<span>  <span class="va">treeML_fixed</span>,</span>
<span>  <span class="va">funchNumbered</span>,</span>
<span>  node.states <span class="op">=</span> <span class="st">"marginal"</span>,</span>
<span>    <span class="co"># Feed in our chosen model</span></span>
<span>  model <span class="op">=</span> <span class="st">"ER"</span>, </span>
<span>    <span class="co"># The number of rate categories</span></span>
<span>  rate.cat <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    <span class="co"># A rate matrix (for the number of trait factors) to optimise</span></span>
<span>  rate.mat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="cn">NA</span><span class="op">)</span>,<span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; You specified 'fixed.nodes=FALSE' but included a phy object with node labels. These node labels have been removed.</span></span>
<span><span class="co">#&gt; Warning in corHMM::corHMM(treeML_fixed, funchNumbered,</span></span>
<span><span class="co">#&gt; node.states = "marginal", : Branch lengths of 0 detected.</span></span>
<span><span class="co">#&gt; Adding 1e-5 to these branches.</span></span>
<span><span class="co">#&gt; State distribution in data:</span></span>
<span><span class="co">#&gt; States:  1   2   </span></span>
<span><span class="co">#&gt; Counts:  3   9   </span></span>
<span><span class="co">#&gt; Beginning thorough optimization search -- performing 0 random restarts </span></span>
<span><span class="co">#&gt; Finished. Inferring ancestral states using marginal reconstruction.</span></span>
<span></span>
<span>  <span class="co"># Have a look at the output</span></span>
<span><span class="va">cor_margRecon</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Fit</span></span>
<span><span class="co">#&gt;       -lnL      AIC     AICc Rate.cat ntax</span></span>
<span><span class="co">#&gt;  -7.238929 16.47786 16.87786        1   12</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Legend</span></span>
<span><span class="co">#&gt;   1   2 </span></span>
<span><span class="co">#&gt; "0" "1" </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Rates</span></span>
<span><span class="co">#&gt;        (1,R1) (2,R1)</span></span>
<span><span class="co">#&gt; (1,R1)     NA    100</span></span>
<span><span class="co">#&gt; (2,R1)    100     NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Arrived at a reliable solution</span></span>
<span></span>
<span><span class="co"># Choose the colours to use once</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"darkorange"</span>, <span class="st">"darkgreen"</span><span class="op">)</span>,</span>
<span>                 <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Factor-class.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">finchSummary</span><span class="op">$</span><span class="va">beakCat</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span><span class="co"># Let's plot the tree.</span></span>
<span><span class="fu">phytools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phytools/man/plotTree.datamatrix.html" class="external-link">plotTree.datamatrix</a></span><span class="op">(</span><span class="va">treeML_fixed</span>,</span>
<span>                              <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">beakVector</span><span class="op">)</span>,</span>
<span>                              colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">cols</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">FALSE</span>, fsize <span class="op">=</span> <span class="fl">0.45</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"bottomleft"</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Factor-class.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">finchSummary</span><span class="op">$</span><span class="va">beakCat</span><span class="op">)</span>, pch <span class="op">=</span> <span class="fl">22</span>, pt.cex <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>       pt.bg <span class="op">=</span> <span class="va">cols</span>, bty <span class="op">=</span> <span class="st">"n"</span>, cex <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span>
<span><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/nodelabels.html" class="external-link">nodelabels</a></span><span class="op">(</span>pie <span class="op">=</span> <span class="va">cor_margRecon</span><span class="op">$</span><span class="va">states</span>, piecol <span class="op">=</span> <span class="va">cols</span>, cex <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="PhyloPrac4_files/figure-html/marginalReconstruct-1.png" width="672"></p>
<p>That’s a pretty cool plot, I think. It shows the probability at each
node that the ancestral state there was large or small. It also shows
the state in the present day! You can see that, in this case, the
analysis has struggled to give us strong answers. But, closer to those
nodes with more consistent states, you can see that the probabilities
are higher that the ancestor had that state. Maybe it would be better
with more states, or better genetic data that gave us longer branch
lengths, I can’t say for sure!</p>
<div class="alert alert-warning" role="alert">
<blockquote>
<p><strong>Q5:</strong> Is it more likely that the common ancestor of
<em>Camarhynchus pallidus</em> and <em>Geospiza scandens</em> had a
large or a small beak?</p>
</blockquote>
</div>
<div class="alert alert-warning" role="alert">
<blockquote>
<p><strong>Q6</strong> How would you change the colour of the categories
in your phylogeny?</p>
</blockquote>
</div>
<p><strong>A small byword:</strong> Running analyses in almost any
statistical program involves transforming your data into different
formats. Depending on the data and the analyses that you’re doing this
load can be bigger or smaller. However, doing all of these things in a
single program and a single script is really very convenient. I don’t
expect you to remember or understand all of my data manipulations but,
trust me, keep these pracs with you and the code may very well come in
handy for the future. <em>This is how I got started with R.</em></p>
</div>
</div>
<div class="section level2" number="5">
<h2 id="whats-coming-next-week">
<span class="header-section-number">5</span> What’s coming next
week?<a class="anchor" aria-label="anchor" href="#whats-coming-next-week"></a>
</h2>
<p>Next week we will be talking about your assignment and providing help
with the first component of your <strong>Grant Proposal</strong>
assignment. Don’t forget to make sure that you have read and understood
what you need to do for that element of your assignment and the marking
criteria. Most of the marks are assigned just for including the correct
elements — note that this means you could, if you wanted, even use the
data that we have harvested in our pracs (but I’d prefer it if you at
least asked a different question).</p>
<p>We will be available to help talk you through hypotheses and
analyses. It will be a great opportunity to get support on the first
component of your assignment and a head start on the following
component.</p>
<p><strong>Be certain to save your R script from this week to help you
with your assignment!</strong></p>
</div>
<div class="section level2" number="6">
<h2 id="a-reminder-of-the-assignment-guidelines">
<span class="header-section-number">6</span> A reminder of the
assignment guidelines<a class="anchor" aria-label="anchor" href="#a-reminder-of-the-assignment-guidelines"></a>
</h2>
<p><strong>Please check Moodle for the most-recent version</strong></p>
<div class="section level3" number="6.1">
<h3 id="part-a-preliminary-phylogeny-5">
<span class="header-section-number">6.1</span> Part (a) —
preliminary phylogeny (5%)<a class="anchor" aria-label="anchor" href="#part-a-preliminary-phylogeny-5"></a>
</h3>
<p>The idea of this component of the grant proposal assignment is to (i)
demonstrate your data and phylogenetic abilities and (ii) to ensure that
you have preliminary results to describe in your grant application.</p>
<ol style="list-style-type: decimal">
<li>
<p>Build a phylogeny using public data from a clade (group of
organisms) of interest</p>
<ol style="list-style-type: lower-alpha">
<li><p>I would suggest that between 7 and 30 taxa should be used,
however, more might be reasonable. These tree tips could be species or
individuals within a species, so long as you have original data for each
tip.</p></li>
<li><p>Your tree must have an outgroup (probably from a closely related
genus; however, the outgroup does not need to have trait data)</p></li>
</ol>
</li>
<li>
<p>Using trait data from the literature or data extracted from
photos for that clade model that trait along the phylogeny (from the
above point) and summarise how that trait has changed</p>
<ol style="list-style-type: lower-alpha">
<li><p>Your traits could come from a published paper directly, or could
be extracted from photos of your species (e.g., something simple like
beak colour or measurements if photos with scales are available for all
species or individuals)</p></li>
<li><p>You must, in writing and figures, explain how your trait(s) have
evolved across the phylogeny. If your data allow it, consider including
a statistical test.</p></li>
<li><p>Provide references to the (i) genbank accession numbers for the
sequences (or similar), (ii) the original research articles (or similar)
where the data were retrieved or derived, and (iii) your code — the
latter will help demonstrators help you with your code</p></li>
</ol>
</li>
</ol>
<p><strong>Table 1</strong>: A description of each assignment component
and what is expected to achieve a particular score for that component
(total possible mark sums to five)</p>
<table class="table">
<colgroup>
<col width="20%">
<col width="34%">
<col width="24%">
<col width="20%">
</colgroup>
<thead><tr class="header">
<th><strong>C.</strong></th>
<th><strong>1</strong></th>
<th><strong>0.5</strong></th>
<th><strong>0</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>1a</strong></td>
<td>A phylogenetic tree of appropriate size (&gt;7 and &lt;30 tips,
excluding outgroup) is provided</td>
<td>A phylogenetic tree of inappropriate size (&lt;7 or &gt;30 tips
excluding outgroup) is provided</td>
<td>No phylogeny is provided</td>
</tr>
<tr class="even">
<td><strong>1b</strong></td>
<td>An appropriate (e.g., relative within the same family or subfamily)
outgroup is provided</td>
<td>An inappropriate (very distant or internal relative) outgroup is
provided</td>
<td>No outgroup is provided</td>
</tr>
<tr class="odd">
<td><strong>2a</strong></td>
<td>Traits for all tree tips (excluding outgroup) are included and
inferred along the phylogeny</td>
<td>Traits for only some tree tips (excluding outgroup) are provided
and/or not inferred along phylogeny</td>
<td>No traits are provided</td>
</tr>
<tr class="even">
<td><strong>2b</strong></td>
<td>Trait evolution across the phylogeny is well described. For example,
and if relevant to their data, the student describes how and where the
trait has changed (within which clades or species); how that trait might
be related to another (if another was measured); the range, mean, or
median values of that trait across the phylogeny or in particular
groups; etc.</td>
<td>Trait evolution across the phylogeny is poorly thought out and
explained. The student misses major patterns in the phylogeny or
describes them particularly poorly. A clever student might describe
their patterns to their demonstrator or lab mates in one of the pracs
prior to submission</td>
<td>No description of trait evolution is provided</td>
</tr>
<tr class="odd">
<td><strong>2c</strong></td>
<td>(i) Genbank accession numbers, (ii) references to trait sources, and
(iii) student R code are all provided</td>
<td>Only 1–2 of the references (see left) are provided</td>
<td>None of the references are provided</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3" number="6.2">
<h3 id="part-b-project-quality-and-innovation-15">
<span class="header-section-number">6.2</span> Part (b) — Project
quality and innovation (15%)<a class="anchor" aria-label="anchor" href="#part-b-project-quality-and-innovation-15"></a>
</h3>
<p>The idea of this component of the grant proposal assignment is to (i)
identify a knowledge gap and develop a well-structured and thought out
hypothesis, (ii) understand and describe the novelty of your question,
(iii) describe your methods and their appropriateness, and (iv) explain
the importance of your project to a grant agency. We will keep in mind
that the scope of your grant must be smaller than one that you would
submit to ARC or to another granting agency. This means that we are
interested in seeing the quality of your reasoning and writing.</p>
<ol style="list-style-type: decimal">
<li>
<p>Very brief literature background of project (no more than one
paragraph)</p>
<ol style="list-style-type: lower-alpha">
<li>Generally, a good background will start with a broader issue and
narrow down to a more specific one. You won’t have too much space here
so try and be succinct and hit a few important points to get the reader
to where they need to be rather efficiently</li>
</ol>
</li>
<li>
<p>Identifying gap in knowledge</p>
<ol style="list-style-type: lower-alpha">
<li>Identify the knowledge gap that you are filling — please note that
this will probably be a minor gap given the scope of the assignment, and
that’s fine!</li>
</ol>
</li>
<li>
<p>Lay out your hypothesis, which is informed by background
knowledge/theory (<strong>not a research question!</strong>)</p>
<ol style="list-style-type: lower-alpha">
<li>Provide a hypothesis that will help you test and address the
knowledge gap. Hypotheses, rather than research questions, are
explicitly testable (i.e. an expected and alternative outcome is clear).
For example, “<em>The relationship between A and B will be C</em>”. Or,
from our pracs was “<em>Finches with larger beak heights will have
larger beak height ranges than finches with smaller beak heights</em>”.
This is compared with a research question like “<em>We will examine how
finch beak height impacts beak height range</em>”</li>
</ol>
</li>
<li>
<p>Novelty/originality and innovation of the proposed research (has
this question been asked before? How is your project novel)</p>
<ol style="list-style-type: lower-alpha">
<li>Explain why this research hasn’t been done before and how it’s new
(I don’t expect you to do an exhaustive literature search, but do
provide some citations for related research here). This can be hard to
show but maybe you can find a novel angle or ecological tint to your
hypothesis.</li>
</ol>
</li>
<li>
<p>Appropriateness of the proposed research design (how will your
project answer the hypothesis that you’re wanting to address)</p>
<ol style="list-style-type: lower-alpha">
<li>Given the context that you have included from points 1 and 2 about,
explain how you will answer your <em>hypothesis</em> using your
phylogenetic methods. NOTE: referring back to your pilot phylogeny could
be very convincing!</li>
</ol>
</li>
<li>
<p>Potential to create new knowledge and research capacity
(blue-sky), and economic, commercial, environmental, social and/or
cultural benefits for Australia (societal benefits).</p>
<ol style="list-style-type: lower-alpha">
<li><p>What is the potential for your project to result in new knowledge
or new research? You could discuss how your work will lead to new
research (follow up research), provide new and important understandings,
expand the field or bring it up to speed with better known taxa (e.g.,
birds are very well known, whereas reptiles and invertebrates are less
well known), etc. (blue-sky component)</p></li>
<li><p>Why is your project important? This could be the importance for
government, conservation, commercial applications (e.g., agricultural),
or cultural benefits for Australia (societal benefits component). For
many ecological questions, conservation can be a safe option to fall
back on</p></li>
</ol>
</li>
</ol>
<p><strong>Table 2</strong>: A description of each assignment component
and what is expected to achieve a particular score for that component
(total possible mark sums to fifteen)</p>
<table class="table">
<colgroup>
<col width="19%">
<col width="21%">
<col width="20%">
<col width="19%">
<col width="19%">
</colgroup>
<thead><tr class="header">
<th><strong>C.</strong></th>
<th><strong>2.5</strong></th>
<th><strong>1.25</strong></th>
<th><strong>0.5</strong></th>
<th><strong>0</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>1a</strong></td>
<td>The background is clear and succinct, bringing the reader from the
broader topic to the specific gap and hypothesis professionally</td>
<td>The background is somewhat clear but inefficient, still brings the
reader from the broader topic to the specific gap and hypothesis</td>
<td>The background if rambling and unclear, the topics discussed are of
dubious relevance but guide the reader to the gap and hypothesis</td>
<td>The background is absent or completely irrelevant so that the reader
is not at all prepared for the specific gap and hypothesis</td>
</tr>
<tr class="even">
<td><strong>2a</strong></td>
<td>NA</td>
<td>Student clearly and succinctly identifies a knowledge gap of strong
relevance to their project and methods</td>
<td>Student identifies a knowledge gap of dubious relevance to their
project and methods</td>
<td>Student does not identify and state a knowledge gap</td>
</tr>
<tr class="odd">
<td><strong>3a</strong></td>
<td>NA</td>
<td>Student clearly and succinctly and provides a testable
hypothesis</td>
<td>Student only provides a research question</td>
<td>Student does not provide a hypothesis or research question</td>
</tr>
<tr class="even">
<td><strong>4a</strong></td>
<td>Student identifies clearly how the research is novel, and provides
references (2–5) to support their argument</td>
<td>Student identifies how their research is novel but may not be clear.
The student provides references (2–5) to support their argument</td>
<td>Student poorly very identifies how their research is novel and
probably provides insufficient references (0–1)</td>
<td>Student does not identify how their research is novel and provides
no references</td>
</tr>
<tr class="odd">
<td><strong>5a</strong></td>
<td>Provides a clear argument that is supported by the sections above
and references (at least 2) as to why their methods are appropriate and
will address the hypothesis. Phylogeny is used well to justify their
methods</td>
<td>Provides an argument that may not be well supported by the above
sections or references but that it is clear how they will address the
hypothesis. Phylogeny is used to justify their methods</td>
<td>Provides poorly-reasoned arguments or sufficient references but it
is clear how they will address their hypothesis. Phylogeny is used to
justify their methods</td>
<td>Provides no arguments for how their methods will address the
hypothesis</td>
</tr>
<tr class="even">
<td><strong>6a</strong></td>
<td>Student provides a clear, thoughtful, and succinct explanation of
the potential for new knowledge or new research resulting from their
proposal that is also justified with references</td>
<td>Student provides some explanation of the potential for new knowledge
or new research resulting from their proposal that is also justified
with references</td>
<td>Student provides a poor explanation of the potential for new
knowledge or new research resulting from their proposal that might not
be justified with references</td>
<td>The students don’t offer any explanation of the potential for new
knowledge or new research resulting from their proposal or
references</td>
</tr>
<tr class="odd">
<td><strong>6b</strong></td>
<td>Student clearly and thoughtfully explains the importance of the
project for one or more of the mentioned bodies/themes (government,
conservation…) such that the reader is convinced that the project is
worth funding</td>
<td>Student provides arguments for the importance of the project for one
or more of the mentioned bodies/themes (government, conservation…) but
the reader is left with serious doubts about funding the project</td>
<td>Arguments for arguments for the importance of the project for one or
more of the mentioned bodies/themes (government, conservation…) but it
is unclear that the project would have any benefits</td>
<td>No arguments are provided.</td>
</tr>
</tbody>
</table>
<p><br><br><br><br></p>
</div>
</div>
<div class="section level2" number="7">
<h2 id="new-packages-used-today">
<span class="header-section-number">7</span> New packages used
today<a class="anchor" aria-label="anchor" href="#new-packages-used-today"></a>
</h2>
<p><img src="Media_EP4/readxl.png" width="200"><img src="Media_EP4/phytools.png" width="200"> And <strong>corHMM</strong>
:/</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James B. Dorey.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
